{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# pycoQC CLI Usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "PycoQC CLI can generate a beautiful HTML formatted report containing interactive D3.js plots. On top of it, the CLI can also dump summary information in a JSON formated file allowing easy parsing with third party tools.\n",
    "\n",
    "The report is dynamically generated depending on the information available in the summary file."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CLI Usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Activate virtual environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-26T15:00:02.927028Z",
     "start_time": "2019-09-26T15:00:02.151389Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(pycoQC) (pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Using virtualenvwrapper here but can also be done with Conda \n",
    "workon pycoQC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Getting help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-26T15:00:07.082866Z",
     "start_time": "2019-09-26T15:00:04.869169Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: pycoQC [-h] [--version]\n",
      "              [--summary_file [SUMMARY_FILE [SUMMARY_FILE ...]]]\n",
      "              [--barcode_file [BARCODE_FILE [BARCODE_FILE ...]]]\n",
      "              [--bam_file [BAM_FILE [BAM_FILE ...]]]\n",
      "              [--html_outfile HTML_OUTFILE] [--json_outfile JSON_OUTFILE]\n",
      "              [--min_pass_qual MIN_PASS_QUAL] [--filter_calibration]\n",
      "              [--filter_duplicated]\n",
      "              [--min_barcode_percent MIN_BARCODE_PERCENT]\n",
      "              [--report_title REPORT_TITLE] [--template_file TEMPLATE_FILE]\n",
      "              [--config_file CONFIG_FILE] [--sample SAMPLE] [--default_config]\n",
      "              [-v | -q]\n",
      "\n",
      "pycoQC computes metrics and generates interactive QC plots from the sequencing summary report generated by Oxford Nanopore technologies basecallers\n",
      "\n",
      "* Minimal usage\n",
      "    pycoQC -f sequencing_summary.txt -o pycoQC_output.html\n",
      "* Including Guppy barcoding file + html output + json output + log output\n",
      "    pycoQC -f sequencing_summary.txt -b barcoding_sequencing.txt -o pycoQC_output.html -j pycoQC_output.json -g pycoQC.log\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "  --version             show program's version number and exit\n",
      "  -v, --verbose         Increase verbosity\n",
      "  -q, --quiet           Reduce verbosity\n",
      "\n",
      "Input/output options:\n",
      "  --summary_file [SUMMARY_FILE [SUMMARY_FILE ...]], -f [SUMMARY_FILE [SUMMARY_FILE ...]]\n",
      "                        Path to a sequencing_summary generated by Albacore\n",
      "                        1.0.0 + (read_fast5_basecaller.py) / Guppy 2.1.3+\n",
      "                        (guppy_basecaller). One can also pass multiple space\n",
      "                        separated file paths or a UNIX style regex matching\n",
      "                        multiple files (Required)\n",
      "  --barcode_file [BARCODE_FILE [BARCODE_FILE ...]], -b [BARCODE_FILE [BARCODE_FILE ...]]\n",
      "                        Path to the barcode_file generated by Guppy 2.1.3+\n",
      "                        (guppy_barcoder) or Deepbinner 0.2.0+. This is not a\n",
      "                        required file. One can also pass multiple space\n",
      "                        separated file paths or a UNIX style regex matching\n",
      "                        multiple files (optional)\n",
      "  --bam_file [BAM_FILE [BAM_FILE ...]], -a [BAM_FILE [BAM_FILE ...]]\n",
      "                        Path to a Bam file corresponding to reads in the\n",
      "                        summary_file. Preferably aligned with Minimap2 One can\n",
      "                        also pass multiple space separated file paths or a\n",
      "                        UNIX style regex matching multiple files (optional)\n",
      "                        NOT IMPLEMENTED YET\n",
      "  --html_outfile HTML_OUTFILE, -o HTML_OUTFILE\n",
      "                        Path to an output html file report (required if\n",
      "                        json_outfile not given)\n",
      "  --json_outfile JSON_OUTFILE, -j JSON_OUTFILE\n",
      "                        Path to an output json file report (required if\n",
      "                        html_outfile not given)\n",
      "\n",
      "Filtering options:\n",
      "  --min_pass_qual MIN_PASS_QUAL\n",
      "                        Minimum quality to consider a read as 'pass' (default:\n",
      "                        7)\n",
      "  --filter_calibration  If given, reads flagged as calibration strand by the\n",
      "                        basecaller are removed (default: False)\n",
      "  --filter_duplicated   If given, duplicated read_ids are removed but the\n",
      "                        first occurence is kept (Guppy sometimes outputs the\n",
      "                        same read multiple times) (default: False)\n",
      "  --min_barcode_percent MIN_BARCODE_PERCENT\n",
      "                        Minimal percent of total reads to retain barcode\n",
      "                        label. If below, the barcode value is set as\n",
      "                        `unclassified` (default: 0.1)\n",
      "\n",
      "HTML report options:\n",
      "  --report_title REPORT_TITLE\n",
      "                        Title to use in the html report (default: PycoQC\n",
      "                        report)\n",
      "  --template_file TEMPLATE_FILE\n",
      "                        Jinja2 html template for the html report (default: )\n",
      "  --config_file CONFIG_FILE\n",
      "                        Path to a JSON configuration file for the html report.\n",
      "                        If not provided, looks for it in ~/.pycoQC and\n",
      "                        ~/.config/pycoQC/config. If it's still not found,\n",
      "                        falls back to default parameters. The first level keys\n",
      "                        are the names of the plots to be included. The second\n",
      "                        level keys are the parameters to pass to each plotting\n",
      "                        function (default: )\")\n",
      "\n",
      "Other options:\n",
      "  --sample SAMPLE       If not None a n number of reads will be randomly\n",
      "                        selected instead of the entire dataset for ploting\n",
      "                        function (deterministic sampling) (default: 100000)\n",
      "  --default_config, -d  Print default configuration file. Can be used to\n",
      "                        generate a template JSON file (default: False)\n",
      "(pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoQC -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Usage examples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Basic usage  (quiet mode)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-26T15:03:35.385494Z",
     "start_time": "2019-09-26T15:03:28.179046Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking arguments values\n",
      "Check input data files\n",
      "Parse data files\n",
      "Merge data\n",
      "Cleaning data\n",
      "Loading plotting interface\n",
      "(pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoQC \\\n",
    "    -f ./data/Albacore-1.2.1_basecall-1D-DNA_sequencing_summary.txt.gz \\\n",
    "    -o ./results/pycoQC/Albacore-1.2.1_basecall-1D-DNA.html \\\n",
    "    --quiet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[HTML OUTPUT](https://a-slide.github.io/pycoQC/demo/results/pycoQC/Albacore-1.2.1_basecall-1D-DNA.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### JSON data output on top of the html report\n",
    "\n",
    "A json report can be generated on top (or instead) of the html report\n",
    "\n",
    "It contains a summarized version of the data collected by pycoQC in a structured and easy to parse format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-26T15:03:47.894952Z",
     "start_time": "2019-09-26T15:03:41.013206Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking arguments values\n",
      "Check input data files\n",
      "Parse data files\n",
      "Merge data\n",
      "Cleaning data\n",
      "\tDiscarding lines containing NA values\n",
      "\t\t0 reads discarded\n",
      "\tFiltering out zero length reads\n",
      "\t\t0 reads discarded\n",
      "\tSorting run IDs by decreasing throughput\n",
      "\t\tRun-id order ['9835d20f1d205bdbd1fb4d464ae778de95beab24']\n",
      "\tReordering runids\n",
      "\t\tProcessing reads with Run_ID 9835d20f1d205bdbd1fb4d464ae778de95beab24 / time offset: 0\n",
      "\tCast value to appropriate type\n",
      "\tReindexing dataframe by read_ids\n",
      "\t\t10,000 Final valid reads\n",
      "Loading plotting interface\n",
      "Generating HTML report\n",
      "\tParsing html config file\n",
      "\tRunning method summary\n",
      "\t\tComputing plot\n",
      "\tRunning method barcode_summary\n",
      "\t\tNo barcode information available\n",
      "\tRunning method run_id_summary\n",
      "\t\tThere is only one run_id\n",
      "\tRunning method read_len_1D\n",
      "\t\tComputing plot\n",
      "\tRunning method align_len_1D\n",
      "\t\tNo Alignment information available\n",
      "\tRunning method read_qual_1D\n",
      "\t\tComputing plot\n",
      "\tRunning method align_score_1D\n",
      "\t\tNo align score information available\n",
      "\tRunning method read_len_read_qual_2D\n",
      "\t\tComputing plot\n",
      "\tRunning method read_len_align_len_2D\n",
      "\t\tNo Alignment information available\n",
      "\tRunning method align_len_align_score_2D\n",
      "\t\tNo align score information available\n",
      "\tRunning method read_qual_align_score_2D\n",
      "\t\tNo align score information available\n",
      "\tRunning method output_over_time\n",
      "\t\tComputing plot\n",
      "\tRunning method len_over_time\n",
      "\t\tComputing plot\n",
      "\tRunning method qual_over_time\n",
      "\t\tComputing plot\n",
      "\tRunning method barcode_counts\n",
      "\t\tNo barcode information available\n",
      "\tRunning method channels_activity\n",
      "\t\tComputing plot\n",
      "\tRunning method alignment_summary\n",
      "\t\tNo Alignment information available\n",
      "\tRunning method alignment_rate\n",
      "\t\tNo align score information available\n",
      "\tRunning method alignment_coverage\n",
      "\t\tNo Alignment information available\n",
      "\tLoading HTML template\n",
      "\tRendering plots in d3js\n",
      "\tWriting to HTML file\n",
      "Generating JSON report\n",
      "\tRunning summary_stats_dict method\n",
      "\tCompute overall summary statistics\n",
      "\tWriting to JSON file\n",
      "(pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoQC \\\n",
    "    -f ./data/Guppy-2.1.3_basecall-1D-RNA_sequencing_summary.txt.gz \\\n",
    "    -o ./results/pycoQC/Guppy-2.1.3_basecall-1D_RNA.html \\\n",
    "    -j ./results/pycoQC/Guppy-2.1.3_basecall-1D_RNA.json"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[HTML OUTPUT](https://a-slide.github.io/pycoQC/demo/results/pycoQC/Guppy-2.1.3_basecall-1D_RNA.html)\n",
    "\n",
    "[JSON OUTPUT](https://a-slide.github.io/pycoQC/demo/results/pycoQC/Guppy-2.1.3_basecall-1D_RNA.json)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Including guppy barcoding information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-26T15:04:30.744875Z",
     "start_time": "2019-09-26T15:04:23.387974Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking arguments values\n",
      "Check input data files\n",
      "Parse data files\n",
      "Merge data\n",
      "Cleaning data\n",
      "\tDiscarding lines containing NA values\n",
      "\t\t0 reads discarded\n",
      "\tFiltering out zero length reads\n",
      "\t\t0 reads discarded\n",
      "\tSorting run IDs by decreasing throughput\n",
      "\t\tRun-id order ['c4981b897c2bb47fed99916c19c9bd1bd43267a2']\n",
      "\tReordering runids\n",
      "\t\tProcessing reads with Run_ID c4981b897c2bb47fed99916c19c9bd1bd43267a2 / time offset: 0\n",
      "\tCleaning up low frequency barcodes\n",
      "\t\t0 reads with low frequency barcode unset\n",
      "\tCast value to appropriate type\n",
      "\tReindexing dataframe by read_ids\n",
      "\t\t10,000 Final valid reads\n",
      "Loading plotting interface\n",
      "Generating HTML report\n",
      "\tParsing html config file\n",
      "\tRunning method summary\n",
      "\t\tComputing plot\n",
      "\tRunning method barcode_summary\n",
      "\t\tComputing plot\n",
      "\tRunning method run_id_summary\n",
      "\t\tThere is only one run_id\n",
      "\tRunning method read_len_1D\n",
      "\t\tComputing plot\n",
      "\tRunning method align_len_1D\n",
      "\t\tNo Alignment information available\n",
      "\tRunning method read_qual_1D\n",
      "\t\tComputing plot\n",
      "\tRunning method align_score_1D\n",
      "\t\tNo align score information available\n",
      "\tRunning method read_len_read_qual_2D\n",
      "\t\tComputing plot\n",
      "\tRunning method read_len_align_len_2D\n",
      "\t\tNo Alignment information available\n",
      "\tRunning method align_len_align_score_2D\n",
      "\t\tNo align score information available\n",
      "\tRunning method read_qual_align_score_2D\n",
      "\t\tNo align score information available\n",
      "\tRunning method output_over_time\n",
      "\t\tComputing plot\n",
      "\tRunning method len_over_time\n",
      "\t\tComputing plot\n",
      "\tRunning method qual_over_time\n",
      "\t\tComputing plot\n",
      "\tRunning method barcode_counts\n",
      "\t\tComputing plot\n",
      "\tRunning method channels_activity\n",
      "\t\tComputing plot\n",
      "\tRunning method alignment_summary\n",
      "\t\tNo Alignment information available\n",
      "\tRunning method alignment_rate\n",
      "\t\tNo align score information available\n",
      "\tRunning method alignment_coverage\n",
      "\t\tNo Alignment information available\n",
      "\tLoading HTML template\n",
      "\tRendering plots in d3js\n",
      "\tWriting to HTML file\n",
      "(pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoQC \\\n",
    "    -f ./data/Guppy-2.1.3_basecall-1D-DNA_sequencing_summary.txt.gz \\\n",
    "    -b ./data/Guppy-2.1.3_basecall-1D_DNA_barcoding_summary.txt.gz \\\n",
    "    -o ./results/pycoQC/Guppy-2.1.3_basecall-1D_DNA_barcode.html"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[HTML OUTPUT](https://a-slide.github.io/pycoQC/demo/results/pycoQC/Guppy-2.1.3_basecall-1D_DNA_barcode.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Matching multiple files with a regex and add a title to report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-26T15:08:48.746358Z",
     "start_time": "2019-09-26T15:08:40.649985Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking arguments values\n",
      "Check input data files\n",
      "Parse data files\n",
      "Merge data\n",
      "Cleaning data\n",
      "\tDiscarding lines containing NA values\n",
      "\t\t0 reads discarded\n",
      "\tFiltering out zero length reads\n",
      "\t\t813 reads discarded\n",
      "\tSorting run IDs by decreasing throughput\n",
      "\t\tRun-id order ['7ae4f0a6d2b7ba3e0248496b7de9cd5d1c028415', '5074e0cd71f372314c30ca5158aab2172d915023', 'c675730269f2f96f300f1cfa613fe89c53b344c3', '2b9163100702bba6ac29d37dbc96ccad740aa05d', 'd0054681152930b21276405d948b115e46968ca6', '71055637dd56eca9416305332eba1ed37bbfffe1', '9835d20f1d205bdbd1fb4d464ae778de95beab24', 'db5916f2fe7957afac1d0aaccdec883342c4bc31', '93fa1ad3ebc8a6e505d991bcb052c2b8ceb278b5', '17b317b994031430f350cda1dc13a72f66572ece']\n",
      "\tReordering runids\n",
      "\t\tProcessing reads with Run_ID 7ae4f0a6d2b7ba3e0248496b7de9cd5d1c028415 / time offset: 0\n",
      "\t\tProcessing reads with Run_ID 5074e0cd71f372314c30ca5158aab2172d915023 / time offset: 5309.74734\n",
      "\t\tProcessing reads with Run_ID c675730269f2f96f300f1cfa613fe89c53b344c3 / time offset: 15911.26726\n",
      "\t\tProcessing reads with Run_ID 2b9163100702bba6ac29d37dbc96ccad740aa05d / time offset: 16306.31175\n",
      "\t\tProcessing reads with Run_ID d0054681152930b21276405d948b115e46968ca6 / time offset: 16694.80113\n",
      "\t\tProcessing reads with Run_ID 71055637dd56eca9416305332eba1ed37bbfffe1 / time offset: 17090.52523\n",
      "\t\tProcessing reads with Run_ID 9835d20f1d205bdbd1fb4d464ae778de95beab24 / time offset: 61285.92364\n",
      "\t\tProcessing reads with Run_ID db5916f2fe7957afac1d0aaccdec883342c4bc31 / time offset: 220697.58266999997\n",
      "\t\tProcessing reads with Run_ID 93fa1ad3ebc8a6e505d991bcb052c2b8ceb278b5 / time offset: 393486.23117\n",
      "\t\tProcessing reads with Run_ID 17b317b994031430f350cda1dc13a72f66572ece / time offset: 393850.01437\n",
      "\tCast value to appropriate type\n",
      "\tReindexing dataframe by read_ids\n",
      "\t\t29,187 Final valid reads\n",
      "Loading plotting interface\n",
      "Generating HTML report\n",
      "\tParsing html config file\n",
      "\tRunning method summary\n",
      "\t\tComputing plot\n",
      "\tRunning method barcode_summary\n",
      "\t\tNo barcode information available\n",
      "\tRunning method run_id_summary\n",
      "\t\tComputing plot\n",
      "\tRunning method read_len_1D\n",
      "\t\tComputing plot\n",
      "\tRunning method align_len_1D\n",
      "\t\tNo Alignment information available\n",
      "\tRunning method read_qual_1D\n",
      "\t\tComputing plot\n",
      "\tRunning method align_score_1D\n",
      "\t\tNo align score information available\n",
      "\tRunning method read_len_read_qual_2D\n",
      "\t\tComputing plot\n",
      "\tRunning method read_len_align_len_2D\n",
      "\t\tNo Alignment information available\n",
      "\tRunning method align_len_align_score_2D\n",
      "\t\tNo align score information available\n",
      "\tRunning method read_qual_align_score_2D\n",
      "\t\tNo align score information available\n",
      "\tRunning method output_over_time\n",
      "\t\tComputing plot\n",
      "\tRunning method len_over_time\n",
      "\t\tComputing plot\n",
      "\tRunning method qual_over_time\n",
      "\t\tComputing plot\n",
      "\tRunning method barcode_counts\n",
      "\t\tNo barcode information available\n",
      "\tRunning method channels_activity\n",
      "\t\tComputing plot\n",
      "\tRunning method alignment_summary\n",
      "\t\tNo Alignment information available\n",
      "\tRunning method alignment_rate\n",
      "\t\tNo align score information available\n",
      "\tRunning method alignment_coverage\n",
      "\t\tNo Alignment information available\n",
      "\tLoading HTML template\n",
      "\tRendering plots in d3js\n",
      "\tWriting to HTML file\n",
      "(pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoQC \\\n",
    "    -f ./data/Albacore*RNA* \\\n",
    "    -o ./results/pycoQC/Albacore_all_RNA.html \\\n",
    "    --report_title \"All RNA runs\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[HTML OUTPUT](https://a-slide.github.io/pycoQC/demo/results/pycoQC/Albacore_all_RNA.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Tweak filtering parameters"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Flag reads with a quality above 8 as pass\n",
    "* Discard reads aligned on the calibration standard\n",
    "* Unset value of any barcode found in less than 10% of the reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-26T15:09:10.602444Z",
     "start_time": "2019-09-26T15:09:03.159075Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking arguments values\n",
      "Check input data files\n",
      "Parse data files\n",
      "Merge data\n",
      "Cleaning data\n",
      "\tDiscarding lines containing NA values\n",
      "\t\t0 reads discarded\n",
      "\tFiltering out zero length reads\n",
      "\t\t40 reads discarded\n",
      "\tFiltering out calibration strand reads\n",
      "\t\t0 reads discarded\n",
      "\tSorting run IDs by decreasing throughput\n",
      "\t\tRun-id order ['aae4df85078f7fe690547aeb688f2640644f323c', 'f54aa9064eb703797b98c83804bd65541b1ffc1b']\n",
      "\tReordering runids\n",
      "\t\tProcessing reads with Run_ID aae4df85078f7fe690547aeb688f2640644f323c / time offset: 0\n",
      "\t\tProcessing reads with Run_ID f54aa9064eb703797b98c83804bd65541b1ffc1b / time offset: 72.23357\n",
      "\tCast value to appropriate type\n",
      "\tReindexing dataframe by read_ids\n",
      "\t\t9,960 Final valid reads\n",
      "Loading plotting interface\n",
      "Generating HTML report\n",
      "\tParsing html config file\n",
      "\tRunning method summary\n",
      "\t\tComputing plot\n",
      "\tRunning method barcode_summary\n",
      "\t\tNo barcode information available\n",
      "\tRunning method run_id_summary\n",
      "\t\tComputing plot\n",
      "\tRunning method read_len_1D\n",
      "\t\tComputing plot\n",
      "\tRunning method align_len_1D\n",
      "\t\tNo Alignment information available\n",
      "\tRunning method read_qual_1D\n",
      "\t\tComputing plot\n",
      "\tRunning method align_score_1D\n",
      "\t\tNo align score information available\n",
      "\tRunning method read_len_read_qual_2D\n",
      "\t\tComputing plot\n",
      "\tRunning method read_len_align_len_2D\n",
      "\t\tNo Alignment information available\n",
      "\tRunning method align_len_align_score_2D\n",
      "\t\tNo align score information available\n",
      "\tRunning method read_qual_align_score_2D\n",
      "\t\tNo align score information available\n",
      "\tRunning method output_over_time\n",
      "\t\tComputing plot\n",
      "\tRunning method len_over_time\n",
      "\t\tComputing plot\n",
      "\tRunning method qual_over_time\n",
      "\t\tComputing plot\n",
      "\tRunning method barcode_counts\n",
      "\t\tNo barcode information available\n",
      "\tRunning method channels_activity\n",
      "\t\tComputing plot\n",
      "\tRunning method alignment_summary\n",
      "\t\tNo Alignment information available\n",
      "\tRunning method alignment_rate\n",
      "\t\tNo align score information available\n",
      "\tRunning method alignment_coverage\n",
      "\t\tNo Alignment information available\n",
      "\tLoading HTML template\n",
      "\tRendering plots in d3js\n",
      "\tWriting to HTML file\n",
      "(pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoQC \\\n",
    "    -f ./data/Albacore-2.1.10_basecall-1D-DNA_sequencing_summary.txt.gz\\\n",
    "    -o ./results/pycoQC/Albacore-2.1.10_basecall-1D-DNA.html \\\n",
    "    --min_pass_qual 8 \\\n",
    "    --filter_calibration \\\n",
    "    --min_barcode_percent 10"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[HTML OUTPUT](https://a-slide.github.io/pycoQC/demo/results/pycoQC/Albacore-2.1.10_basecall-1D-DNA.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Including Alignments information for a Bam file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-26T15:09:57.855067Z",
     "start_time": "2019-09-26T15:09:46.830141Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking arguments values\n",
      "Check input data files\n",
      "Parse data files\n",
      "Merge data\n",
      "Cleaning data\n",
      "\tDiscarding lines containing NA values\n",
      "\t\t0 reads discarded\n",
      "\tFiltering out zero length reads\n",
      "\t\t0 reads discarded\n",
      "\tSorting run IDs by decreasing throughput\n",
      "\t\tRun-id order ['ede7e01619570f500a43fd5d33ff8ab25d1b589b']\n",
      "\tReordering runids\n",
      "\t\tProcessing reads with Run_ID ede7e01619570f500a43fd5d33ff8ab25d1b589b / time offset: 0\n",
      "\tCast value to appropriate type\n",
      "\tReindexing dataframe by read_ids\n",
      "\t\t12,000 Final valid reads\n",
      "Loading plotting interface\n",
      "Generating HTML report\n",
      "\tParsing html config file\n",
      "\tRunning method summary\n",
      "\t\tComputing plot\n",
      "\tRunning method barcode_summary\n",
      "\t\tNo barcode information available\n",
      "\tRunning method run_id_summary\n",
      "\t\tThere is only one run_id\n",
      "\tRunning method read_len_1D\n",
      "\t\tComputing plot\n",
      "\tRunning method align_len_1D\n",
      "\t\tComputing plot\n",
      "\tRunning method read_qual_1D\n",
      "\t\tComputing plot\n",
      "\tRunning method align_score_1D\n",
      "\t\tComputing plot\n",
      "\tRunning method read_len_read_qual_2D\n",
      "\t\tComputing plot\n",
      "\tRunning method read_len_align_len_2D\n",
      "\t\tComputing plot\n",
      "\tRunning method align_len_align_score_2D\n",
      "\t\tComputing plot\n",
      "\tRunning method read_qual_align_score_2D\n",
      "\t\tComputing plot\n",
      "\tRunning method output_over_time\n",
      "\t\tComputing plot\n",
      "\tRunning method len_over_time\n",
      "\t\tComputing plot\n",
      "\tRunning method qual_over_time\n",
      "\t\tComputing plot\n",
      "\tRunning method barcode_counts\n",
      "\t\tNo barcode information available\n",
      "\tRunning method channels_activity\n",
      "\t\tComputing plot\n",
      "\tRunning method alignment_summary\n",
      "\t\tComputing plot\n",
      "\tRunning method alignment_rate\n",
      "\t\tComputing plot\n",
      "\tRunning method alignment_coverage\n",
      "\t\tComputing plot\n",
      "\tLoading HTML template\n",
      "\tRendering plots in d3js\n",
      "\tWriting to HTML file\n",
      "Generating JSON report\n",
      "\tRunning summary_stats_dict method\n",
      "\tCompute overall summary statistics\n",
      "\tWriting to JSON file\n",
      "(pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoQC \\\n",
    "    -f ./large_data/sample_1_sequencing_summary.txt \\\n",
    "    -a ./large_data/sample_1.bam \\\n",
    "    -o ./results/pycoQC/Guppy-2.3_basecall-1D_alignment-DNA.html \\\n",
    "    -j ./results/pycoQC/Guppy-2.3_basecall-1D_alignment-DNA.json"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[HTML OUTPUT](https://a-slide.github.io/pycoQC/demo/results/pycoQC/Guppy-2.3_basecall-1D_alignment-DNA.html)\n",
    "\n",
    "[JSON OUTPUT](https://a-slide.github.io/pycoQC/demo/results/pycoQC/Guppy-2.3_basecall-1D_alignment-DNA.json)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Advanced configuration with custon json file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Although we recommend to stick to the default parameters, a json formatted configuration file can be provided to tweak the plots. A default configuration file can be generated using:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-26T15:11:13.567033Z",
     "start_time": "2019-09-26T15:11:11.307014Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"summary\": {\n",
      "    \"plot_title\": \"Run summary\"\n",
      "  },\n",
      "  \"barcode_summary\": {\n",
      "    \"plot_title\": \"Run summary by barcode\"\n",
      "  },\n",
      "  \"run_id_summary\": {\n",
      "    \"plot_title\": \"Run summary by Run ID\"\n",
      "  },\n",
      "  \"read_len_1D\": {\n",
      "    \"plot_title\": \"Basecalled reads length\",\n",
      "    \"color\": \"lightsteelblue\",\n",
      "    \"nbins\": 200,\n",
      "    \"smooth_sigma\": 2\n",
      "  },\n",
      "  \"align_len_1D\": {\n",
      "    \"plot_title\": \"Reads alignment length\",\n",
      "    \"color\": \"mediumseagreen\",\n",
      "    \"nbins\": 200,\n",
      "    \"smooth_sigma\": 2\n",
      "  },\n",
      "  \"read_qual_1D\": {\n",
      "    \"plot_title\": \"Reads PHRED quality\",\n",
      "    \"color\": \"salmon\",\n",
      "    \"nbins\": 200,\n",
      "    \"smooth_sigma\": 2\n",
      "  },\n",
      "  \"align_score_1D\": {\n",
      "    \"plot_title\": \"Reads alignment score\",\n",
      "    \"color\": \"sandybrown\",\n",
      "    \"nbins\": 200,\n",
      "    \"smooth_sigma\": 2\n",
      "  },\n",
      "  \"read_len_read_qual_2D\": {\n",
      "    \"plot_title\": \"Basecalled reads length vs reads PHRED quality\",\n",
      "    \"x_nbins\": 200,\n",
      "    \"y_nbins\": 100,\n",
      "    \"smooth_sigma\": 2\n",
      "  },\n",
      "  \"read_len_align_len_2D\": {\n",
      "    \"plot_title\": \"Basecalled reads length vs alignments length\",\n",
      "    \"x_nbins\": 200,\n",
      "    \"y_nbins\": 100,\n",
      "    \"smooth_sigma\": 1\n",
      "  },\n",
      "  \"align_len_align_score_2D\": {\n",
      "    \"plot_title\": \"Aligned reads length vs alignments score\",\n",
      "    \"x_nbins\": 200,\n",
      "    \"y_nbins\": 100,\n",
      "    \"smooth_sigma\": 2\n",
      "  },\n",
      "  \"read_qual_align_score_2D\": {\n",
      "    \"plot_title\": \"Mean read quality per sequence length\",\n",
      "    \"x_nbins\": 200,\n",
      "    \"y_nbins\": 100,\n",
      "    \"smooth_sigma\": 1\n",
      "  },\n",
      "  \"output_over_time\": {\n",
      "    \"plot_title\": \"Output over experiment time\",\n",
      "    \"cumulative_color\": \"rgb(204,226,255)\",\n",
      "    \"interval_color\": \"rgb(102,168,255)\"\n",
      "  },\n",
      "  \"len_over_time\": {\n",
      "    \"plot_title\": \"Read length over time\",\n",
      "    \"median_color\": \"rgb(102,168,255)\",\n",
      "    \"quartile_color\": \"rgb(153,197,255)\",\n",
      "    \"extreme_color\": \"rgba(153,197,255,0.5)\",\n",
      "    \"smooth_sigma\": 1\n",
      "  },\n",
      "  \"qual_over_time\": {\n",
      "    \"plot_title\": \"Mean read quality over time\",\n",
      "    \"median_color\": \"rgb(250,128,114)\",\n",
      "    \"quartile_color\": \"rgb(250,170,160)\",\n",
      "    \"extreme_color\": \"rgba(250,170,160,0.5)\",\n",
      "    \"smooth_sigma\": 1\n",
      "  },\n",
      "  \"barcode_counts\": {\n",
      "    \"plot_title\": \"Number of reads per barcode\",\n",
      "    \"colors\": [\n",
      "      \"#f8bc9c\",\n",
      "      \"#f6e9a1\",\n",
      "      \"#f5f8f2\",\n",
      "      \"#92d9f5\",\n",
      "      \"#4f97ba\"\n",
      "    ]\n",
      "  },\n",
      "  \"channels_activity\": {\n",
      "    \"plot_title\": \"Channel activity over time\",\n",
      "    \"smooth_sigma\": 1\n",
      "  },\n",
      "  \"alignment_summary\": {\n",
      "    \"plot_title\": \"Summary of reads alignment\",\n",
      "    \"colors\": [\n",
      "      \"#f8bc9c\",\n",
      "      \"#f6e9a1\",\n",
      "      \"#92d9f5\",\n",
      "      \"#4f97ba\",\n",
      "      \"#f5f8f2\"\n",
      "    ]\n",
      "  },\n",
      "  \"alignment_rate\": {\n",
      "    \"plot_title\": \"Bases alignment rate breakdown\",\n",
      "    \"colors\": [\n",
      "      \"#fcaf94\",\n",
      "      \"#828282\",\n",
      "      \"#fc8161\",\n",
      "      \"#828282\",\n",
      "      \"#f44f39\",\n",
      "      \"#d52221\",\n",
      "      \"#828282\",\n",
      "      \"#828282\",\n",
      "      \"#828282\",\n",
      "      \"#828282\"\n",
      "    ]\n",
      "  },\n",
      "  \"alignment_coverage\": {\n",
      "    \"plot_title\": \"Coverage overview\",\n",
      "    \"nbins\": 500,\n",
      "    \"color\": \"rgba(102,168,255,0.75)\",\n",
      "    \"smooth_sigma\": 1\n",
      "  }\n",
      "}\n",
      "(pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoQC --default_config"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-06-24T15:52:37.096001Z",
     "start_time": "2019-06-24T15:52:36.888372Z"
    }
   },
   "source": [
    "To save and edit it redirect the std output to a file and make your changes using your favorite text editor.\n",
    "\n",
    "To remove a plot from the report, just remove it (or comment it) from the configuration file\n",
    "\n",
    "The configuration file accept all the arguments of the target plotting functions. For more information refer to the API documentation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-26T13:31:19.511766Z",
     "start_time": "2019-09-26T13:31:18.773Z"
    }
   },
   "outputs": [],
   "source": [
    "pycoQC --default_config > data/pycoQC_config.json"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run pycoQC with `--config` option"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-26T15:11:39.914741Z",
     "start_time": "2019-09-26T15:11:33.590718Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking arguments values\n",
      "Check input data files\n",
      "Parse data files\n",
      "Merge data\n",
      "Cleaning data\n",
      "\tDiscarding lines containing NA values\n",
      "\t\t0 reads discarded\n",
      "\tFiltering out zero length reads\n",
      "\t\t438 reads discarded\n",
      "\tSorting run IDs by decreasing throughput\n",
      "\t\tRun-id order ['db5916f2fe7957afac1d0aaccdec883342c4bc31']\n",
      "\tReordering runids\n",
      "\t\tProcessing reads with Run_ID db5916f2fe7957afac1d0aaccdec883342c4bc31 / time offset: 0\n",
      "\tCleaning up low frequency barcodes\n",
      "\t\t0 reads with low frequency barcode unset\n",
      "\tCast value to appropriate type\n",
      "\tReindexing dataframe by read_ids\n",
      "\t\t9,562 Final valid reads\n",
      "Loading plotting interface\n",
      "Generating HTML report\n",
      "\tParsing html config file\n",
      "\tRunning method summary\n",
      "\t\tComputing plot\n",
      "\tRunning method read_len_1D\n",
      "\t\tComputing plot\n",
      "\tRunning method read_qual_1D\n",
      "\t\tComputing plot\n",
      "\tRunning method read_len_read_qual_2D\n",
      "\t\tComputing plot\n",
      "\tRunning method output_over_time\n",
      "\t\tComputing plot\n",
      "\tRunning method len_over_time\n",
      "\t\tComputing plot\n",
      "\tRunning method qual_over_time\n",
      "\t\tComputing plot\n",
      "\tLoading HTML template\n",
      "\tRendering plots in d3js\n",
      "\tWriting to HTML file\n",
      "(pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoQC \\\n",
    "    -f ./data/Albacore-1.7.0_basecall-1D-DNA_sequencing_summary.txt.gz\\\n",
    "    -o ./results/pycoQC/Albacore-1.7.0_basecall-1D-DNA.html \\\n",
    "    --config ./data/pycoQC_config.json"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[HTML OUTPUT](https://a-slide.github.io/pycoQC/demo/results/pycoQC/Albacore-1.7.0_basecall-1D-DNA.html)"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Initialization Cell",
  "hide_input": false,
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  },
  "widgets": {
   "state": {},
   "version": "1.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
