{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# pycoQC CLI Usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "PycoQC CLI can generate a beautiful HTML formatted report containing interactive D3.js plots. On top of it, the CLI can also dump summary information in a JSON formated file allowing easy parsing with third party tools.\n",
    "\n",
    "The report is dynamically generated depending on the information available in the summary file."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CLI Usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Activate virtual environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-07T13:28:08.401253Z",
     "start_time": "2019-05-07T13:28:07.611467Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(pycoQC) (pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Using virtualenvwrapper here but can also be done with Conda \n",
    "workon pycoQC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Getting help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-07T13:28:10.711678Z",
     "start_time": "2019-05-07T13:28:09.611134Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: pycoQC [-h] [--version]\n",
      "              [--summary_file SUMMARY_FILE [SUMMARY_FILE ...]]\n",
      "              [--barcode_file BARCODE_FILE [BARCODE_FILE ...]]\n",
      "              [--html_outfile HTML_OUTFILE] [--json_outfile JSON_OUTFILE]\n",
      "              [--min_pass_qual MIN_PASS_QUAL] [--filter_calibration]\n",
      "              [--min_barcode_percent MIN_BARCODE_PERCENT] [--title TITLE]\n",
      "              [--template_file TEMPLATE_FILE] [--config CONFIG]\n",
      "              [--default_config] [--list_plots] [--verbose_level {2,1,0}]\n",
      "\n",
      "pycoQC computes metrics and generates interactive QC plots from the sequencing summary report generated by Oxford Nanopore technologies basecallers\n",
      "\n",
      "* Minimal usage\n",
      "    pycoQC -f sequencing_summary.txt -o pycoQC_output.html\n",
      "* Including Guppy barcoding file and json output\n",
      "    pycoQC -f sequencing_summary.txt -b barcoding_sequencing.txt -o pycoQC_output.html -j pycoQC_output.json\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "  --version             show program's version number and exit\n",
      "\n",
      "Input/output options:\n",
      "  --summary_file SUMMARY_FILE [SUMMARY_FILE ...], -f SUMMARY_FILE [SUMMARY_FILE ...]\n",
      "                        Path to the sequencing_summary generated by Albacore\n",
      "                        1.0.0 + (read_fast5_basecaller.py) / Guppy 2.1.3+\n",
      "                        (guppy_basecaller). One can also pass multiple space\n",
      "                        separated file paths or a UNIX style regex matching\n",
      "                        multiple files (Required)\n",
      "  --barcode_file BARCODE_FILE [BARCODE_FILE ...], -b BARCODE_FILE [BARCODE_FILE ...]\n",
      "                        Path to the barcode_summary_file generated by Guppy\n",
      "                        2.1.3+ (guppy_barcoder).. One can also pass multiple\n",
      "                        space separated file paths or a UNIX style regex\n",
      "                        matching multiple files (optional)\n",
      "  --html_outfile HTML_OUTFILE, -o HTML_OUTFILE\n",
      "                        Path to an output html file (required if json_outfile\n",
      "                        not given)\n",
      "  --json_outfile JSON_OUTFILE, -j JSON_OUTFILE\n",
      "                        Path to an output json file (required if html_outfile\n",
      "                        not given)\n",
      "\n",
      "Filtering options:\n",
      "  --min_pass_qual MIN_PASS_QUAL, -q MIN_PASS_QUAL\n",
      "                        Minimum quality to consider a read as 'pass' (default:\n",
      "                        7)\n",
      "  --filter_calibration  If given reads flagged as calibration strand by the\n",
      "                        basecaller are removed (default: False)\n",
      "  --min_barcode_percent MIN_BARCODE_PERCENT\n",
      "                        Minimal percent of total reads to retain barcode\n",
      "                        label. If below the barcode value is set as\n",
      "                        `unclassified` (default: 0.1)\n",
      "\n",
      "HTML report options:\n",
      "  --title TITLE, -t TITLE\n",
      "                        A title to be used in the html report (default: None)\n",
      "  --template_file TEMPLATE_FILE\n",
      "                        Jinja2 html template for the html report (default:\n",
      "                        None)\n",
      "  --config CONFIG, -c CONFIG\n",
      "                        Path to a JSON configuration file for the html report.\n",
      "                        If not provided, looks for it in ~/.pycoQC and\n",
      "                        ~/.config/pycoQC/config. If it's still not found,\n",
      "                        falls back to default parameters. The first level keys\n",
      "                        are the names of the plots to be included. The second\n",
      "                        level keys are the parameters to pass to each plotting\n",
      "                        function (default: None)\")\n",
      "\n",
      "Other options:\n",
      "  --default_config, -d  Print default configuration file. Can be used to\n",
      "                        generate a template JSON file (default: False)\n",
      "  --list_plots, -l      Print the list of available plotting functions and\n",
      "                        exit (default: None)\n",
      "  --verbose_level {2,1,0}, -v {2,1,0}\n",
      "                        Level of verbosity, from 2 (Chatty) to 0 (Nothing)\n",
      "                        (default: 1)\n",
      "(pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoQC -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Usage examples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Basic usage "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-07T13:28:18.989473Z",
     "start_time": "2019-05-07T13:28:14.233642Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PARSING DATA FILES\n",
      "Importing raw data from sequencing summary files\n",
      "Verifying fields and discarding unused columns\n",
      "Droping lines containing NA values\n",
      "Sorting run IDs by decreasing throughput\n",
      "Reordering runids\n",
      "Reindexing dataframe by read_ids\n",
      "GENERATING HTML REPORT\n",
      "\tParsing html config file\n",
      "\tRunning method summary\n",
      "\tRunning method barcode_summary\n",
      "\t\tNo barcode information available\n",
      "\tRunning method run_id_summary\n",
      "\tRunning method reads_len_1D\n",
      "\tRunning method reads_qual_1D\n",
      "\tRunning method reads_len_qual_2D\n",
      "\tRunning method output_over_time\n",
      "\tRunning method len_over_time\n",
      "\tRunning method qual_over_time\n",
      "\tRunning method barcode_counts\n",
      "\t\tNo barcode information available\n",
      "\tRunning method channels_activity\n",
      "\tLoading HTML template\n",
      "\tRendering plots in d3js\n",
      "\tWriting to HTML file\n",
      "(pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoQC \\\n",
    "    -f data/summary/Albacore-1.2.1_basecall-1D-DNA_sequencing_summary.txt.gz \\\n",
    "    -o data/output/Albacore-1.2.1_basecall-1D-DNA.html"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[HTML OUTPUT](https://a-slide.github.io/pycoQC/demo/data/output/Albacore-1.2.1_basecall-1D-DNA.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### JSON data output on top of the html report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-07T13:28:24.369114Z",
     "start_time": "2019-05-07T13:28:19.738692Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PARSING DATA FILES\n",
      "Importing raw data from sequencing summary files\n",
      "Verifying fields and discarding unused columns\n",
      "Droping lines containing NA values\n",
      "Sorting run IDs by decreasing throughput\n",
      "Reordering runids\n",
      "Reindexing dataframe by read_ids\n",
      "GENERATING HTML REPORT\n",
      "\tParsing html config file\n",
      "\tRunning method summary\n",
      "\tRunning method barcode_summary\n",
      "\t\tNo barcode information available\n",
      "\tRunning method run_id_summary\n",
      "\tRunning method reads_len_1D\n",
      "\tRunning method reads_qual_1D\n",
      "\tRunning method reads_len_qual_2D\n",
      "\tRunning method output_over_time\n",
      "\tRunning method len_over_time\n",
      "\tRunning method qual_over_time\n",
      "\tRunning method barcode_counts\n",
      "\t\tNo barcode information available\n",
      "\tRunning method channels_activity\n",
      "\tLoading HTML template\n",
      "\tRendering plots in d3js\n",
      "\tWriting to HTML file\n",
      "GENERATING JSON REPORT\n",
      "\tRunning summary_stats_dict method\n",
      "\tWriting to JSON file\n",
      "(pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoQC \\\n",
    "    -f data/summary/Guppy-2.1.3_basecall-1D-RNA_sequencing_summary.txt.gz \\\n",
    "    -o data/output/Guppy-2.1.3_basecall-1D_RNA.html \\\n",
    "    -j data/output/Guppy-2.1.3_basecall-1D_RNA.json"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[HTML OUTPUT](https://a-slide.github.io/pycoQC/demo/data/output/Guppy-2.1.3_basecall-1D_RNA.html)\n",
    "\n",
    "[JSON OUTPUT](https://a-slide.github.io/pycoQC/demo/data/output/Guppy-2.1.3_basecall-1D_RNA.json)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Including guppy barcoding information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-07T13:28:39.321875Z",
     "start_time": "2019-05-07T13:28:34.672632Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PARSING DATA FILES\n",
      "Importing raw data from sequencing summary files\n",
      "Importing barcode information from barcode summary files\n",
      "Verifying fields and discarding unused columns\n",
      "Droping lines containing NA values\n",
      "Sorting run IDs by decreasing throughput\n",
      "Reordering runids\n",
      "Cleaning up low frequency barcodes\n",
      "Reindexing dataframe by read_ids\n",
      "GENERATING HTML REPORT\n",
      "\tParsing html config file\n",
      "\tRunning method summary\n",
      "\tRunning method barcode_summary\n",
      "\tRunning method run_id_summary\n",
      "\tRunning method reads_len_1D\n",
      "\tRunning method reads_qual_1D\n",
      "\tRunning method reads_len_qual_2D\n",
      "\tRunning method output_over_time\n",
      "\tRunning method len_over_time\n",
      "\tRunning method qual_over_time\n",
      "\tRunning method barcode_counts\n",
      "\tRunning method channels_activity\n",
      "\tLoading HTML template\n",
      "\tRendering plots in d3js\n",
      "\tWriting to HTML file\n",
      "(pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoQC \\\n",
    "    -f data/summary/Guppy-2.1.3_basecall-1D-DNA_sequencing_summary.txt.gz \\\n",
    "    -b data/summary/Guppy-2.1.3_basecall-1D_DNA_barcoding_summary.txt.gz \\\n",
    "    -o data/output/Guppy-2.1.3_basecall-1D_DNA_barcode.html"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[HTML OUTPUT](https://a-slide.github.io/pycoQC/demo/data/output/Guppy-2.1.3_basecall-1D_DNA_barcode.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Matching multiple files with a regex and add a title to report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-07T13:28:30.576778Z",
     "start_time": "2019-05-07T13:28:25.202798Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PARSING DATA FILES\n",
      "Importing raw data from sequencing summary files\n",
      "Verifying fields and discarding unused columns\n",
      "Droping lines containing NA values\n",
      "Filtering out zero length reads\n",
      "Sorting run IDs by decreasing throughput\n",
      "Reordering runids\n",
      "Reindexing dataframe by read_ids\n",
      "GENERATING HTML REPORT\n",
      "\tParsing html config file\n",
      "\tRunning method summary\n",
      "\tRunning method barcode_summary\n",
      "\t\tNo barcode information available\n",
      "\tRunning method run_id_summary\n",
      "\tRunning method reads_len_1D\n",
      "\tRunning method reads_qual_1D\n",
      "\tRunning method reads_len_qual_2D\n",
      "\tRunning method output_over_time\n",
      "\tRunning method len_over_time\n",
      "\tRunning method qual_over_time\n",
      "\tRunning method barcode_counts\n",
      "\t\tNo barcode information available\n",
      "\tRunning method channels_activity\n",
      "\tLoading HTML template\n",
      "\tRendering plots in d3js\n",
      "\tWriting to HTML file\n",
      "(pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoQC \\\n",
    "    -f data/summary/Albacore*RNA* \\\n",
    "    -o data/output/Albacore_all_RNA.html \\\n",
    "    --title \"All RNA runs\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[HTML OUTPUT](https://a-slide.github.io/pycoQC/demo/data/output/Albacore_all_RNA.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Tweak filtering parameters"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Flag reads with a quality above 8 as pass\n",
    "* Discard reads aligned on the calibration standard\n",
    "* Unset value of any barcode found in less than 10% of the reads\n",
    "* Increase verbose_level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-07T13:28:46.652877Z",
     "start_time": "2019-05-07T13:28:42.110382Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PARSING DATA FILES\n",
      "Importing raw data from sequencing summary files\n",
      "\tSequencing summary files found: ['data/summary/Albacore-2.1.10_basecall-1D-DNA_sequencing_summary.txt.gz']\n",
      "\t10,000 reads found in initial file\n",
      "Verifying fields and discarding unused columns\n",
      "\t1D Run type\n",
      "\tColumns found: ['read_id', 'run_id', 'channel', 'start_time', 'sequence_length_template', 'mean_qscore_template', 'calibration_strand_genome_template']\n",
      "Droping lines containing NA values\n",
      "\t0 reads discarded\n",
      "Filtering out zero length reads\n",
      "\t40 reads discarded\n",
      "Filtering out calibration strand reads\n",
      "\t0 reads discarded\n",
      "Sorting run IDs by decreasing throughput\n",
      "\tRun-id order ['aae4df85078f7fe690547aeb688f2640644f323c', 'f54aa9064eb703797b98c83804bd65541b1ffc1b']\n",
      "Reordering runids\n",
      "\tProcessing reads with Run_ID aae4df85078f7fe690547aeb688f2640644f323c / time offset: 0\n",
      "\tProcessing reads with Run_ID f54aa9064eb703797b98c83804bd65541b1ffc1b / time offset: 72.23357\n",
      "Reindexing dataframe by read_ids\n",
      "[pycoQC]\n",
      "Runtime info\n",
      "\tpackage_name: pycoQC\n",
      "\tpackage_version: 2.2.3.7\n",
      "\ttimestamp: 2019-05-07 14:28:43.307422\n",
      "\tverbose_level: 2\n",
      "\tmin_barcode_percent: 10.0\n",
      "\tfilter_calibration: True\n",
      "\tmin_pass_qual: 8\n",
      "\trunid_list: []\n",
      "\tbarcode_summary_file: None\n",
      "\tseq_summary_file: ['data/summary/Albacore-2.1.10_basecall-1D-DNA_sequencing_summary.txt.gz']\n",
      "Read counts\n",
      "\tInitial reads: 10000\n",
      "\tReads with NA values discarded: 0\n",
      "\tZero length reads discarded: 40\n",
      "\tCalibration reads discarded: 0\n",
      "\tValid reads: 9960\n",
      "\tValid pass reads: 2535\n",
      "\n",
      "GENERATING HTML REPORT\n",
      "\tParsing html config file\n",
      "\tTry to read config file from system directory:/home/aleg/.pycoQC\n",
      "\t\tFile not found, non-readable or invalid\n",
      "\tTry to read config file from system directory:/home/aleg/.config/pycoQC/config\n",
      "\t\tFile not found, non-readable or invalid\n",
      "\tFall back to default configuration\n",
      "{'summary': {'plot_title': 'Run summary'}, 'barcode_summary': {'plot_title': 'Run summary by barcode'}, 'run_id_summary': {'plot_title': 'Run summary by Run ID'}, 'reads_len_1D': {'plot_title': 'Distribution of read length', 'color': 'lightsteelblue', 'nbins': 200, 'smooth_sigma': 2, 'sample': 100000}, 'reads_qual_1D': {'plot_title': 'Distribution of read quality', 'color': 'salmon', 'nbins': 200, 'smooth_sigma': 2, 'sample': 100000}, 'reads_len_qual_2D': {'plot_title': 'Mean read quality per sequence length', 'colorscale': [[0.0, 'rgba(255,255,255,0)'], [0.1, 'rgba(255,150,0,0)'], [0.25, 'rgb(255,100,0)'], [0.5, 'rgb(200,0,0)'], [0.75, 'rgb(120,0,0)'], [1.0, 'rgb(70,0,0)']], 'len_nbins': 200, 'qual_nbins': 75, 'smooth_sigma': 2, 'sample': 100000}, 'output_over_time': {'plot_title': 'Output over experiment time', 'cumulative_color': 'rgb(204,226,255)', 'interval_color': 'rgb(102,168,255)', 'sample': 100000}, 'len_over_time': {'plot_title': 'Read length over time', 'median_color': 'rgb(102,168,255)', 'quartile_color': 'rgb(153,197,255)', 'extreme_color': 'rgba(153,197,255,0.5)', 'smooth_sigma': 1, 'sample': 100000}, 'qual_over_time': {'plot_title': 'Mean read quality over time', 'median_color': 'rgb(250,128,114)', 'quartile_color': 'rgb(250,170,160)', 'extreme_color': 'rgba(250,170,160,0.5)', 'smooth_sigma': 1, 'sample': 100000}, 'barcode_counts': {'plot_title': 'Number of reads per barcode', 'colors': ['#f8bc9c', '#f6e9a1', '#f5f8f2', '#92d9f5', '#4f97ba']}, 'channels_activity': {'plot_title': 'Channel activity over time', 'colorscale': [[0.0, 'rgba(255,255,255,0)'], [0.01, 'rgb(255,255,200)'], [0.25, 'rgb(255,200,0)'], [0.5, 'rgb(200,0,0)'], [0.75, 'rgb(120,0,0)'], [1.0, 'rgb(0,0,0)']], 'smooth_sigma': 1, 'sample': 100000}}\n",
      "\tRunning method summary\n",
      "\tsummary ({'plot_title': 'Run summary'})\n",
      "\tRunning method barcode_summary\n",
      "\tbarcode_summary ({'plot_title': 'Run summary by barcode'})\n",
      "\t\tNo barcode information available\n",
      "\tRunning method run_id_summary\n",
      "\trun_id_summary ({'plot_title': 'Run summary by Run ID'})\n",
      "\tRunning method reads_len_1D\n",
      "\treads_len_1D ({'plot_title': 'Distribution of read length', 'color': 'lightsteelblue', 'nbins': 200, 'smooth_sigma': 2, 'sample': 100000})\n",
      "\tRunning method reads_qual_1D\n",
      "\treads_qual_1D ({'plot_title': 'Distribution of read quality', 'color': 'salmon', 'nbins': 200, 'smooth_sigma': 2, 'sample': 100000})\n",
      "\tRunning method reads_len_qual_2D\n",
      "\treads_len_qual_2D ({'plot_title': 'Mean read quality per sequence length', 'colorscale': [[0.0, 'rgba(255,255,255,0)'], [0.1, 'rgba(255,150,0,0)'], [0.25, 'rgb(255,100,0)'], [0.5, 'rgb(200,0,0)'], [0.75, 'rgb(120,0,0)'], [1.0, 'rgb(70,0,0)']], 'len_nbins': 200, 'qual_nbins': 75, 'smooth_sigma': 2, 'sample': 100000})\n",
      "\tRunning method output_over_time\n",
      "\toutput_over_time ({'plot_title': 'Output over experiment time', 'cumulative_color': 'rgb(204,226,255)', 'interval_color': 'rgb(102,168,255)', 'sample': 100000})\n",
      "\tRunning method len_over_time\n",
      "\tlen_over_time ({'plot_title': 'Read length over time', 'median_color': 'rgb(102,168,255)', 'quartile_color': 'rgb(153,197,255)', 'extreme_color': 'rgba(153,197,255,0.5)', 'smooth_sigma': 1, 'sample': 100000})\n",
      "\tRunning method qual_over_time\n",
      "\tqual_over_time ({'plot_title': 'Mean read quality over time', 'median_color': 'rgb(250,128,114)', 'quartile_color': 'rgb(250,170,160)', 'extreme_color': 'rgba(250,170,160,0.5)', 'smooth_sigma': 1, 'sample': 100000})\n",
      "\tRunning method barcode_counts\n",
      "\tbarcode_counts ({'plot_title': 'Number of reads per barcode', 'colors': ['#f8bc9c', '#f6e9a1', '#f5f8f2', '#92d9f5', '#4f97ba']})\n",
      "\t\tNo barcode information available\n",
      "\tRunning method channels_activity\n",
      "\tchannels_activity ({'plot_title': 'Channel activity over time', 'colorscale': [[0.0, 'rgba(255,255,255,0)'], [0.01, 'rgb(255,255,200)'], [0.25, 'rgb(255,200,0)'], [0.5, 'rgb(200,0,0)'], [0.75, 'rgb(120,0,0)'], [1.0, 'rgb(0,0,0)']], 'smooth_sigma': 1, 'sample': 100000})\n",
      "\tLoading HTML template\n",
      "\tFall back to default template\n",
      "\tRendering plots in d3js\n",
      "\tWriting to HTML file\n",
      "(pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoQC \\\n",
    "    -f data/summary/Albacore-2.1.10_basecall-1D-DNA_sequencing_summary.txt.gz\\\n",
    "    -o data/output/Albacore-2.1.10_basecall-1D-DNA.html \\\n",
    "    --min_pass_qual 8 \\\n",
    "    --filter_calibration \\\n",
    "    --min_barcode_percent 10 \\\n",
    "    --verbose_level 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[HTML OUTPUT](https://a-slide.github.io/pycoQC/demo/data/output/Albacore-2.1.10_basecall-1D-DNA.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Advanced configuration with custon json file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Although we recommend to stick to the default parameters, a json formatted configuration file can be provided to tweak the plots. A default configuration file can be generated using:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-07T13:28:54.907845Z",
     "start_time": "2019-05-07T13:28:53.748412Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoQC --default_config > data/pycoQC_config.json"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make changes to the file in your favorite text editor (channels_activity and barcode_counts plots was removed in this case). For more information refer to the API documentation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run pycoQC with `--config` option"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-07T13:29:27.289135Z",
     "start_time": "2019-05-07T13:29:23.408576Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PARSING DATA FILES\n",
      "Importing raw data from sequencing summary files\n",
      "Verifying fields and discarding unused columns\n",
      "Droping lines containing NA values\n",
      "Filtering out zero length reads\n",
      "Sorting run IDs by decreasing throughput\n",
      "Reordering runids\n",
      "Cleaning up low frequency barcodes\n",
      "Reindexing dataframe by read_ids\n",
      "GENERATING HTML REPORT\n",
      "\tParsing html config file\n",
      "\tRunning method summary\n",
      "\tRunning method barcode_summary\n",
      "\tRunning method run_id_summary\n",
      "\tRunning method reads_len_1D\n",
      "\tRunning method reads_qual_1D\n",
      "\tRunning method reads_len_qual_2D\n",
      "\tRunning method output_over_time\n",
      "\tRunning method len_over_time\n",
      "\tRunning method qual_over_time\n",
      "\tLoading HTML template\n",
      "\tRendering plots in d3js\n",
      "\tWriting to HTML file\n",
      "(pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoQC \\\n",
    "    -f data/summary/Albacore-1.7.0_basecall-1D-DNA_sequencing_summary.txt.gz\\\n",
    "    -o data/output/Albacore-1.7.0_basecall-1D-DNA.html \\\n",
    "    --config data/pycoQC_config.json"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[HTML OUTPUT](https://a-slide.github.io/pycoQC/demo/data/output/Albacore-1.7.0_basecall-1D-DNA.html)"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Initialization Cell",
  "hide_input": false,
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  },
  "widgets": {
   "state": {},
   "version": "1.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
