{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# pycoQC API Usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "pycoQC is a simple class that is initialized with a text summary file generated by ONT Albacore or Guppy.\n",
    "\n",
    "The instantiated object can be subsequently called with various methods that will generates tables and plots.\n",
    "\n",
    "There are a few different ways to get help for all the public package functions:\n",
    "\n",
    "* In a separate window with the jupyter magic \"?\": `?pycoQC.channels_activity`\n",
    "* In an output cell with the standard help function: `help (pycoQC.channels_activity)`\n",
    "* Inline with the cursor on the function of interest use `shift` + `tab` "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running pycoQC in Jupyter notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you want to run pycoQC interactively in Jupyter you need to install Jupyter manually. If you installed pycoQC in a virtual environment then install Jupyter in the same virtual environment.\n",
    "\n",
    "```bash\n",
    "pip3 install notebook\n",
    "```\n",
    "\n",
    "Launch the notebook in a shell terminal\n",
    "\n",
    "```bash\n",
    "jupyter notebook\n",
    "```\n",
    "\n",
    "If it does not auto-start, open the following URL in you favorite web browser http://localhost:8888/tree\n",
    "\n",
    "From Jupyter homepage you can navigate to the directory you want to work in and create a new Python3 Notebook.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### For plotly offline plotting"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-11-07T13:18:37.130404Z",
     "start_time": "2018-11-07T13:18:37.126613Z"
    }
   },
   "source": [
    "Import pycoQC main class as well as Plotly and enable inline plotting in the current Notebook.\n",
    "\n",
    "This is the recommended option. This ensures that your all your data are stored inside the notebook.\n",
    "\n",
    "The limitation is that if generating many plots with large datasets the notebook will become quite heavy and slow.   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:24.768214Z",
     "start_time": "2020-02-10T14:23:24.695435Z"
    },
    "init_cell": true
   },
   "outputs": [],
   "source": [
    "# Run cell with Ctrl + Enter in Jupyter\n",
    "\n",
    "# Import main pycoQC module\n",
    "from pycoQC.pycoQC import pycoQC\n",
    "from pycoQC.pycoQC_plot import pycoQC_plot\n",
    "\n",
    "# Import helper functions from pycoQC\n",
    "from pycoQC.common import jhelp\n",
    "\n",
    "# Import and setup plotly for offline plotting in Jupyter \n",
    "from plotly.offline import init_notebook_mode\n",
    "init_notebook_mode (connected=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### For plotly online plotting"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This option takes advantage of Plotly web-service for hosting graphs. This requires to set up an account (https://plot.ly/python/getting-started/#initialization-for-online-plotting) and to provide credentials in the notebook. This could be a good option for easy sharing of the interactive plots generated by pycoQC."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-09T16:48:36.379168Z",
     "start_time": "2020-01-09T16:48:36.376654Z"
    }
   },
   "outputs": [],
   "source": [
    "# Only run this cell if you have set up a plotly account before and wants to use Plotly web-service \n",
    "# from plotly.plotly import plot, iplot\n",
    "# import plotly.tools as pt\n",
    "# pt.set_credentials_file (username=\"XXXXXXXXXX\", api_key=\"XXXXXXXXXX\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initialisation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-10-30T10:45:37.332044Z",
     "start_time": "2018-10-30T10:45:37.323494Z"
    }
   },
   "source": [
    "Upon initialization pycoQC reads the sequencing summary file, runs a series of tests and pre-process the data for plotting methods."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:24.777487Z",
     "start_time": "2020-02-10T14:23:24.772579Z"
    },
    "init_cell": true,
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**pycoQC** (summary_file, barcode_file, bam_file, runid_list, filter_calibration, filter_duplicated, min_barcode_percent, min_pass_qual, min_pass_len, sample, html_outfile, report_title, config_file, template_file, json_outfile, verbose, quiet)\n",
       "\n",
       "Parse Albacore sequencing_summary.txt file and clean-up the data\n",
       "\n",
       "---\n",
       "\n",
       "* **summary_file** (required) [str]\n",
       "\n",
       "Path to a sequencing_summary generated by Albacore 1.0.0 + (read_fast5_basecaller.py) / Guppy 2.1.3+ (guppy_basecaller). One can also pass multiple space separated file paths or a UNIX style regex matching multiple files\n",
       "\n",
       "* **barcode_file** (default: \"\") [str]\n",
       "\n",
       "Path to the barcode_file generated by Guppy 2.1.3+ (guppy_barcoder) or Deepbinner 0.2.0+. This is not a required file. One can also pass multiple space separated file paths or a UNIX style regex matching multiple files\n",
       "\n",
       "* **bam_file** (default: \"\") [str]\n",
       "\n",
       "Path to a Bam file corresponding to reads in the summary_file. Preferably aligned with Minimap2 One can also pass multiple space separated file paths or a UNIX style regex matching multiple files\n",
       "\n",
       "* **runid_list** (default: []) [list]\n",
       "\n",
       "Select only specific runids to be analysed. Can also be used to force pycoQC to order the runids for temporal plots, if the sequencing_summary file contain several sucessive runs. By default pycoQC analyses all the runids in the file and uses the runid order as defined in the file.\n",
       "\n",
       "* **filter_calibration** (default: False) [bool]\n",
       "\n",
       "If True read flagged as calibration strand by the software are removed\n",
       "\n",
       "* **filter_duplicated** (default: False) [bool]\n",
       "\n",
       "If True duplicated read_ids are removed but the first occurence is kept (Guppy sometimes outputs the same read multiple times)\n",
       "\n",
       "* **min_barcode_percent** (default: 0.1) [float]\n",
       "\n",
       "Minimal percent of total reads to retain barcode label. If below the barcode value is set as `unclassified`.\n",
       "\n",
       "* **min_pass_qual** (default: 7) [float]\n",
       "\n",
       "Minimum quality to consider a read as 'pass'\n",
       "\n",
       "* **min_pass_len** (default: 0) [int]\n",
       "\n",
       "Minimum read length to consider a read as 'pass'\n",
       "\n",
       "* **sample** (default: 100000) [int]\n",
       "\n",
       "If not None a n number of reads will be randomly selected instead of the entire dataset for ploting function (deterministic sampling)\n",
       "\n",
       "* **html_outfile** (default: \"\") [str]\n",
       "\n",
       "Path to an output html file report\n",
       "\n",
       "* **report_title** (default: PycoQC report) [str]\n",
       "\n",
       "Title to use in the html report\n",
       "\n",
       "* **config_file** (default: \"\") [str]\n",
       "\n",
       "Path to a JSON configuration file for the html report. If not provided, falls back to default parameters. The first level keys are the names of the plots to be included. The second level keys are the parameters to pass to each plotting function\n",
       "\n",
       "* **template_file** (default: \"\") [str]\n",
       "\n",
       "Jinja2 html template for the html report\n",
       "\n",
       "* **json_outfile** (default: \"\") [str]\n",
       "\n",
       "Path to an output json file report\n",
       "\n",
       "* **verbose** (default: False) [bool]\n",
       "\n",
       "Increase verbosity\n",
       "\n",
       "* **quiet** (default: False) [bool]\n",
       "\n",
       "Reduce verbosity\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp (pycoQC)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Basic initialisation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:50.231041Z",
     "start_time": "2020-02-10T14:23:49.993991Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Checking arguments values\n",
      "Check input data files\n",
      "Parse data files\n",
      "Merge data\n",
      "Cleaning data\n",
      "\tDiscarding lines containing NA values\n",
      "\t\t0 reads discarded\n",
      "\tFiltering out zero length reads\n",
      "\t\t438 reads discarded\n",
      "\tSorting run IDs by decreasing throughput\n",
      "\t\tRun-id order ['db5916f2fe7957afac1d0aaccdec883342c4bc31']\n",
      "\tReordering runids\n",
      "\t\tProcessing reads with Run_ID db5916f2fe7957afac1d0aaccdec883342c4bc31 / time offset: 0\n",
      "\tCleaning up low frequency barcodes\n",
      "\t\t0 reads with low frequency barcode unset\n",
      "\tCast value to appropriate type\n",
      "\tReindexing dataframe by read_ids\n",
      "\t\t9,562 Final valid reads\n",
      "Loading plotting interface\n",
      "\tFound 9,562 total reads\n",
      "\tFound 8,042 pass reads (qual >= 7.0 and length >= 0)\n"
     ]
    }
   ],
   "source": [
    "p = pycoQC(\"./data/Albacore-1.7.0_basecall-1D-DNA_sequencing_summary.txt.gz\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Initialisation with modification of the \"pass\" threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:51.978534Z",
     "start_time": "2020-02-10T14:23:51.746553Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Checking arguments values\n",
      "Check input data files\n",
      "Parse data files\n",
      "Merge data\n",
      "Cleaning data\n",
      "\tDiscarding lines containing NA values\n",
      "\t\t0 reads discarded\n",
      "\tFiltering out zero length reads\n",
      "\t\t438 reads discarded\n",
      "\tSorting run IDs by decreasing throughput\n",
      "\t\tRun-id order ['db5916f2fe7957afac1d0aaccdec883342c4bc31']\n",
      "\tReordering runids\n",
      "\t\tProcessing reads with Run_ID db5916f2fe7957afac1d0aaccdec883342c4bc31 / time offset: 0\n",
      "\tCleaning up low frequency barcodes\n",
      "\t\t0 reads with low frequency barcode unset\n",
      "\tCast value to appropriate type\n",
      "\tReindexing dataframe by read_ids\n",
      "\t\t9,562 Final valid reads\n",
      "Loading plotting interface\n",
      "\tFound 9,562 total reads\n",
      "\tFound 6,693 pass reads (qual >= 8.0 and length >= 200)\n"
     ]
    }
   ],
   "source": [
    "p = pycoQC(\"./data/Albacore-1.7.0_basecall-1D-DNA_sequencing_summary.txt.gz\", min_pass_len=200, min_pass_qual=8)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Initialisation with calibration strand filtering out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:53.073809Z",
     "start_time": "2020-02-10T14:23:52.825144Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Checking arguments values\n",
      "Check input data files\n",
      "Parse data files\n",
      "Merge data\n",
      "Cleaning data\n",
      "\tDiscarding lines containing NA values\n",
      "\t\t0 reads discarded\n",
      "\tFiltering out zero length reads\n",
      "\t\t438 reads discarded\n",
      "\tFiltering out calibration strand reads\n",
      "\t\t2,605 reads discarded\n",
      "\tSorting run IDs by decreasing throughput\n",
      "\t\tRun-id order ['db5916f2fe7957afac1d0aaccdec883342c4bc31']\n",
      "\tReordering runids\n",
      "\t\tProcessing reads with Run_ID db5916f2fe7957afac1d0aaccdec883342c4bc31 / time offset: 0\n",
      "\tCleaning up low frequency barcodes\n",
      "\t\t0 reads with low frequency barcode unset\n",
      "\tCast value to appropriate type\n",
      "\tReindexing dataframe by read_ids\n",
      "\t\t6,957 Final valid reads\n",
      "Loading plotting interface\n",
      "\tFound 6,957 total reads\n",
      "\tFound 5,594 pass reads (qual >= 7.0 and length >= 0)\n"
     ]
    }
   ],
   "source": [
    "p = pycoQC(\"./data/Albacore-1.7.0_basecall-1D-DNA_sequencing_summary.txt.gz\", filter_calibration=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Initialisation with summary file regex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:54.763829Z",
     "start_time": "2020-02-10T14:23:54.181192Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Checking arguments values\n",
      "Check input data files\n",
      "Parse data files\n",
      "Merge data\n",
      "Cleaning data\n",
      "\tDiscarding lines containing NA values\n",
      "\t\t0 reads discarded\n",
      "\tFiltering out zero length reads\n",
      "\t\t813 reads discarded\n",
      "\tSorting run IDs by decreasing throughput\n",
      "\t\tRun-id order ['7ae4f0a6d2b7ba3e0248496b7de9cd5d1c028415', '5074e0cd71f372314c30ca5158aab2172d915023', '9835d20f1d205bdbd1fb4d464ae778de95beab24', 'c675730269f2f96f300f1cfa613fe89c53b344c3', '2b9163100702bba6ac29d37dbc96ccad740aa05d', 'd0054681152930b21276405d948b115e46968ca6', '71055637dd56eca9416305332eba1ed37bbfffe1', 'db5916f2fe7957afac1d0aaccdec883342c4bc31', '93fa1ad3ebc8a6e505d991bcb052c2b8ceb278b5', '17b317b994031430f350cda1dc13a72f66572ece']\n",
      "\tReordering runids\n",
      "\t\tProcessing reads with Run_ID 7ae4f0a6d2b7ba3e0248496b7de9cd5d1c028415 / time offset: 0\n",
      "\t\tProcessing reads with Run_ID 5074e0cd71f372314c30ca5158aab2172d915023 / time offset: 5309.74734\n",
      "\t\tProcessing reads with Run_ID 9835d20f1d205bdbd1fb4d464ae778de95beab24 / time offset: 15911.26726\n",
      "\t\tProcessing reads with Run_ID c675730269f2f96f300f1cfa613fe89c53b344c3 / time offset: 183649.42351\n",
      "\t\tProcessing reads with Run_ID 2b9163100702bba6ac29d37dbc96ccad740aa05d / time offset: 184044.468\n",
      "\t\tProcessing reads with Run_ID d0054681152930b21276405d948b115e46968ca6 / time offset: 184432.95738\n",
      "\t\tProcessing reads with Run_ID 71055637dd56eca9416305332eba1ed37bbfffe1 / time offset: 184828.68148\n",
      "\t\tProcessing reads with Run_ID db5916f2fe7957afac1d0aaccdec883342c4bc31 / time offset: 229024.07989\n",
      "\t\tProcessing reads with Run_ID 93fa1ad3ebc8a6e505d991bcb052c2b8ceb278b5 / time offset: 401812.72839\n",
      "\t\tProcessing reads with Run_ID 17b317b994031430f350cda1dc13a72f66572ece / time offset: 402176.51159\n",
      "\tCast value to appropriate type\n",
      "\tReindexing dataframe by read_ids\n",
      "\t\t39,187 Final valid reads\n",
      "Loading plotting interface\n",
      "\tFound 39,187 total reads\n",
      "\tFound 31,661 pass reads (qual >= 7.0 and length >= 0)\n"
     ]
    }
   ],
   "source": [
    "p = pycoQC(\"./data/*RNA*\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Initialisation with guppy barcoding file and verbose option"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:24:02.362211Z",
     "start_time": "2020-02-10T14:24:02.030196Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Checking arguments values\n",
      "General info\n",
      "\tpackage_name: pycoQC\n",
      "\tpackage_version: 2.5.0.20\n",
      "\ttimestamp: 2020-02-10 14:24:02.031647\n",
      "\n",
      "Runtime options\n",
      "\tquiet: False\n",
      "\tverbose: True\n",
      "\tjson_outfile: \n",
      "\ttemplate_file: \n",
      "\tconfig_file: \n",
      "\treport_title: PycoQC report\n",
      "\thtml_outfile: \n",
      "\tsample: 100000\n",
      "\tmin_pass_len: 0\n",
      "\tmin_pass_qual: 7\n",
      "\tmin_barcode_percent: 0.1\n",
      "\tfilter_duplicated: False\n",
      "\tfilter_calibration: False\n",
      "\trunid_list: []\n",
      "\tbam_file: \n",
      "\tbarcode_file: ./data/Guppy-2.1.3_basecall-1D_DNA_barcoding_summary.txt.gz\n",
      "\tsummary_file: ./data/Guppy-2.1.3_basecall-1D-DNA_sequencing_summary.txt.gz\n",
      "\n",
      "Check input data files\n",
      "\t\tSequencing summary files found: ./data/Guppy-2.1.3_basecall-1D-DNA_sequencing_summary.txt.gz\n",
      "\t\tBarcode files found: ./data/Guppy-2.1.3_basecall-1D_DNA_barcoding_summary.txt.gz\n",
      "Parse data files\n",
      "\tParse summary files\n",
      "\tRename summary sequencing columns\n",
      "\tVerifying fields and discarding unused columns\n",
      "\t\t10,000 reads found in initial file\n",
      "\tParse barcode files\n",
      "\t\tFound valid Guppy barcode file\n",
      "\t\t9,080 reads with barcodes assigned\n",
      "Merge data\n",
      "Cleaning data\n",
      "\tDiscarding lines containing NA values\n",
      "\t\t0 reads discarded\n",
      "\tFiltering out zero length reads\n",
      "\t\t0 reads discarded\n",
      "\tSorting run IDs by decreasing throughput\n",
      "\t\tRun-id order ['c4981b897c2bb47fed99916c19c9bd1bd43267a2']\n",
      "\tReordering runids\n",
      "\t\tProcessing reads with Run_ID c4981b897c2bb47fed99916c19c9bd1bd43267a2 / time offset: 0\n",
      "\tCleaning up low frequency barcodes\n",
      "\t\t0 reads with low frequency barcode unset\n",
      "\tCast value to appropriate type\n",
      "\tReindexing dataframe by read_ids\n",
      "\t\t10,000 Final valid reads\n",
      "Parser stats\n",
      "\tSummary files found: 1\n",
      "\tBarcode files found: 1\n",
      "\tInitial reads: 10000\n",
      "\tReads with barcodes: 9080\n",
      "\tReads with NA values discarded: 0\n",
      "\tZero length reads discarded: 0\n",
      "\tReads with low frequency barcode unset: 0\n",
      "\tValid reads: 10000\n",
      "\n",
      "Loading plotting interface\n",
      "\tFound 10,000 total reads\n",
      "\tFound 9,065 pass reads (qual >= 7.0 and length >= 0)\n",
      "Plotter stats\n",
      "\tBarcode: True\n",
      "\tAlignment: False\n",
      "\tPromethion: False\n",
      "\tAll reads: 10,000\n",
      "\tAll bases: 31,634,280\n",
      "\tAll median read length: 1,223.5\n",
      "\tPass reads: 9,065\n",
      "\tPass bases: 29,973,906\n",
      "\tPass median read length: 1,297.0\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tBarcode: True\n",
      "\tAlignment: False\n",
      "\tPromethion: False\n",
      "\tAll reads: 10,000\n",
      "\tAll bases: 31,634,280\n",
      "\tAll median read length: 1,223.5\n",
      "\tPass reads: 9,065\n",
      "\tPass bases: 29,973,906\n",
      "\tPass median read length: 1,297.0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "p = pycoQC(\n",
    "    summary_file=\"./data/Guppy-2.1.3_basecall-1D-DNA_sequencing_summary.txt.gz\",\n",
    "    barcode_file=\"./data/Guppy-2.1.3_basecall-1D_DNA_barcoding_summary.txt.gz\", verbose=True)\n",
    "print(p)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Initialisation with Deepbinner barcoding file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:24:03.517217Z",
     "start_time": "2020-02-10T14:24:03.337715Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Checking arguments values\n",
      "Check input data files\n",
      "Parse data files\n",
      "Merge data\n",
      "Cleaning data\n",
      "\tDiscarding lines containing NA values\n",
      "\t\t0 reads discarded\n",
      "\tFiltering out zero length reads\n",
      "\t\t0 reads discarded\n",
      "\tSorting run IDs by decreasing throughput\n",
      "\t\tRun-id order ['4318773ae77000ada4c30d67466f7a0462cd5c7a', '13797ccf3e2df77597eb86d8e23088d7f34ba047']\n",
      "\tReordering runids\n",
      "\t\tProcessing reads with Run_ID 4318773ae77000ada4c30d67466f7a0462cd5c7a / time offset: 0\n",
      "\t\tProcessing reads with Run_ID 13797ccf3e2df77597eb86d8e23088d7f34ba047 / time offset: 63.148499\n",
      "\tCleaning up low frequency barcodes\n",
      "\t\t0 reads with low frequency barcode unset\n",
      "\tCast value to appropriate type\n",
      "\tReindexing dataframe by read_ids\n",
      "\t\t3,999 Final valid reads\n",
      "Loading plotting interface\n",
      "\tFound 3,999 total reads\n",
      "\tFound 3,327 pass reads (qual >= 7.0 and length >= 0)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tBarcode: True\n",
      "\tAlignment: False\n",
      "\tPromethion: False\n",
      "\tAll reads: 3,999\n",
      "\tAll bases: 5,690,011\n",
      "\tAll median read length: 914.0\n",
      "\tPass reads: 3,327\n",
      "\tPass bases: 5,007,561\n",
      "\tPass median read length: 961.0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "p = pycoQC(\n",
    "    summary_file=\"./data/Guppy-basecall-1D-DNA_sequencing_summary.txt.gz\",\n",
    "    barcode_file=\"./data/Guppy-basecall-1D-DNA_deepbinner_barcoding_summary.txt.gz\")\n",
    "print(p)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Initialisation with Bam files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:24:07.479154Z",
     "start_time": "2020-02-10T14:24:06.447803Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Checking arguments values\n",
      "Check input data files\n",
      "Parse data files\n",
      "Merge data\n",
      "Cleaning data\n",
      "\tDiscarding lines containing NA values\n",
      "\t\t0 reads discarded\n",
      "\tFiltering out zero length reads\n",
      "\t\t0 reads discarded\n",
      "\tSorting run IDs by decreasing throughput\n",
      "\t\tRun-id order ['ede7e01619570f500a43fd5d33ff8ab25d1b589b']\n",
      "\tReordering runids\n",
      "\t\tProcessing reads with Run_ID ede7e01619570f500a43fd5d33ff8ab25d1b589b / time offset: 0\n",
      "\tCast value to appropriate type\n",
      "\tReindexing dataframe by read_ids\n",
      "\t\t12,000 Final valid reads\n",
      "Loading plotting interface\n",
      "\tFound 12,000 total reads\n",
      "\tFound 10,021 pass reads (qual >= 7.0 and length >= 0)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tBarcode: False\n",
      "\tAlignment: True\n",
      "\tPromethion: False\n",
      "\tAll reads: 12,000\n",
      "\tAll bases: 33,468,903\n",
      "\tAll median read length: 1,011.0\n",
      "\tPass reads: 10,021\n",
      "\tPass bases: 30,288,575\n",
      "\tPass median read length: 1,103.0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "p = pycoQC(\n",
    "    summary_file=\"./large_data/sample_1_sequencing_summary.txt\",\n",
    "    bam_file=\"./large_data/sample_1.bam\")\n",
    "print(p)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generating inline plots and tables"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Interaction with Plotly library"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Most of `pycoQC` methods return a `Figure` object generated with [plotly for Python](https://plot.ly/python/). The `Figure` object can be subsequently used for:\n",
    "\n",
    "* Further customization using the numerous methods attached to the Figure object \n",
    "* Inline plotting in Jupyter Notebook using `iplot` (either from `plotly.plotly` or `plotly.offline`)\n",
    "* Generating a separate HTML file with `plot` (either from `plotly.plotly` or `plotly.offline`)\n",
    "* Exporting as a static image (https://plot.ly/python/static-image-export/), pdf (https://plot.ly/python/pdf-reports/) or various text formats.\n",
    "\n",
    "Users can also customize the figures online in a user friendly environment by clicking on \"Edit in Chart Studio\" in the upper right corner of each figures.\n",
    "\n",
    "![Edit in Chart Studio](https://a-slide.github.io/pycoQC/pictures/edit_in_chart_studio.png)\n",
    "\n",
    "Similarly static pictures can be exported using the \"Download plot as a png\" button.\n",
    "\n",
    "![Edit in Chart Studio](https://a-slide.github.io/pycoQC/pictures/Download_plot_as_png.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Common arguments"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All the methods have the arguments `width` and `height` that can be used to customize the plotting area. In general we do not recommend modifing these values as they might disrupt the plot layout.\n",
    "\n",
    "Most of the methods also have the argument `sample`. By default pycoQC downsample the number of reads to 100,000 before plotting. This drastically reduces the processing time for large dataset and has a very limited impact on the plot aspect. The sampling is random but deterministic, meaning that you should always obtain the same results for the same dataset. The value can be changed to increase or decrease the number of reads. Alternatively, one can deactivate the behavior by specifying `sample=False`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### summary_stats_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`summary_stats_dict` is the only `pycoQC` public method that does not return a plotly `Figure` object. Instead it returns a dictionary containing information about the run options, read counts during the initialisation step and basic statistics for both `pass` and `all` reads.  \n",
    "\n",
    "On top of the overall results, users can also get data split by run_id or barcode.\n",
    "\n",
    "The dictionary can be easily save in a JSON, YAML or pickle file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:24.789268Z",
     "start_time": "2020-02-10T14:23:24.779629Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**summary_stats_dict** ()\n",
       "\n",
       "Return a dictionnary containing exhaustive information about the run.\n",
       "\n",
       "---\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp (pycoQC_plot.summary_stats_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-09T16:48:39.341679Z",
     "start_time": "2020-01-09T16:48:39.107625Z"
    },
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "p = pycoQC(\"./data/Albacore-1.7.0_basecall-1D-DNA_sequencing_summary.txt.gz\", quiet=True)\n",
    "print(p.summary_stats_dict())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### summary table methods"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "pycoQC has 3 summary methods to generate simple summary tables with a clickable button to switch from \"all reads\" to \"pass reads\" only.\n",
    "* `run_summary` : General information about the run\n",
    "* `basecall_summary` : Basecalling related information\n",
    "* `alignment_summary` : Alignment related information (when a bam file is provided"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### run_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:24.802533Z",
     "start_time": "2020-02-10T14:23:24.790960Z"
    },
    "init_cell": true,
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**run_summary** (width, height, plot_title)\n",
       "\n",
       "Plot an interactive overall summary table\n",
       "\n",
       "---\n",
       "\n",
       "* **width** (default: None) [int]\n",
       "\n",
       "With of the plotting area in pixel\n",
       "\n",
       "* **height** (default: 300) [int]\n",
       "\n",
       "height of the plotting area in pixel\n",
       "\n",
       "* **plot_title** (default: General run summary) [str]\n",
       "\n",
       "Title to display on top of the plot\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(pycoQC_plot.run_summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-09T17:13:27.907346Z",
     "start_time": "2020-01-09T17:13:26.679387Z"
    },
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "p = pycoQC(summary_file=\"./data/Guppy-2.2.4-basecall-1D-DNA_sequencing_summary+barcode.txt.gz\", quiet=True)\n",
    "p.run_summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### basecall_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:24.811976Z",
     "start_time": "2020-02-10T14:23:24.804375Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**basecall_summary** (width, height, plot_title)\n",
       "\n",
       "Plot an interactive basecall summary table\n",
       "\n",
       "---\n",
       "\n",
       "* **width** (default: None) [int]\n",
       "\n",
       "With of the plotting area in pixel\n",
       "\n",
       "* **height** (default: 300) [int]\n",
       "\n",
       "height of the plotting area in pixel\n",
       "\n",
       "* **plot_title** (default: Basecall summary) [str]\n",
       "\n",
       "Title to display on top of the plot\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(pycoQC_plot.basecall_summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-09T17:11:07.678232Z",
     "start_time": "2020-01-09T17:11:06.883015Z"
    }
   },
   "outputs": [],
   "source": [
    "p = pycoQC(summary_file=\"./large_data/sample_2_sequencing_summary.txt\", quiet=True)\n",
    "display(p.basecall_summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### alignment_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:24.824125Z",
     "start_time": "2020-02-10T14:23:24.813699Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**alignment_summary** (width, height, plot_title)\n",
       "\n",
       "Plot an interactive alignment summary table\n",
       "\n",
       "---\n",
       "\n",
       "* **width** (default: None) [int]\n",
       "\n",
       "With of the plotting area in pixel\n",
       "\n",
       "* **height** (default: 300) [int]\n",
       "\n",
       "height of the plotting area in pixel\n",
       "\n",
       "* **plot_title** (default: Alignment summary) [str]\n",
       "\n",
       "Title to display on top of the plot\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(pycoQC_plot.alignment_summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-09T16:48:42.547909Z",
     "start_time": "2020-01-09T16:48:41.354743Z"
    }
   },
   "outputs": [],
   "source": [
    "p = pycoQC(summary_file=\"./large_data/sample_2_sequencing_summary.txt\", bam_file=\"./large_data/sample_2.bam\", quiet=True)\n",
    "display(p.alignment_summary())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read Length and Mean quality distribution "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "pycoQC has 3 methods to visualize the distribution of mean quality scores and of estimated read length:\n",
    "* `read_len_1D`: An density plot of estimated read length in logarithmic scale\n",
    "* `read_qual_1D`: An density plot of mean quality scores\n",
    "* `read_len_read_qual_2D`: A density contour plot of estimated read length vs mean quality scores in semilog scale\n",
    "\n",
    "Although we recommend to stick to default values, all 3 methods allow users to customize the plots.\n",
    "* The numbers of bin to divide the reads quality and/or length space in can be specified with `nbins` for the 1D plots and `x_nbins` / `y_nbins` for the 2D plot\n",
    "* The intensity of line smoothing (using a gaussian kernel filter) can be specified with `smooth_sigma`.\n",
    "* Additional cosmetic customization are available: `color`/`colorscale`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:24.834473Z",
     "start_time": "2020-02-10T14:23:24.825886Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**read_len_1D** (color, nbins, smooth_sigma, width, height, plot_title)\n",
       "\n",
       "Plot a distribution of read length (log scale)\n",
       "\n",
       "---\n",
       "\n",
       "* **color** (default: lightsteelblue) [str]\n",
       "\n",
       "Color of the area (hex, rgb, rgba, hsl, hsv or any CSS named colors https://www.w3.org/TR/css-color-3/#svg-color\n",
       "\n",
       "* **nbins** (default: 200) [int]\n",
       "\n",
       "Number of bins to devide the x axis in\n",
       "\n",
       "* **smooth_sigma** (default: 2) [float]\n",
       "\n",
       "standard deviation for Gaussian kernel\n",
       "\n",
       "* **width** (default: None) [int]\n",
       "\n",
       "With of the plotting area in pixel\n",
       "\n",
       "* **height** (default: 500) [int]\n",
       "\n",
       "height of the plotting area in pixel\n",
       "\n",
       "* **plot_title** (default: Basecalled reads length) [str]\n",
       "\n",
       "Title to display on top of the plot\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(pycoQC_plot.read_len_1D)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-09T16:48:43.020731Z",
     "start_time": "2020-01-09T16:48:42.559254Z"
    }
   },
   "outputs": [],
   "source": [
    "p = pycoQC(\"./data/Guppy-2.1.3_basecall-1D-DNA_sequencing_summary.txt.gz\", min_pass_len=200, min_pass_qual=7, quiet=True)\n",
    "p.read_len_1D()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:24.844938Z",
     "start_time": "2020-02-10T14:23:24.836256Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**read_qual_1D** (color, nbins, smooth_sigma, width, height, plot_title)\n",
       "\n",
       "Plot a distribution of quality scores\n",
       "\n",
       "---\n",
       "\n",
       "* **color** (default: salmon) [str]\n",
       "\n",
       "Color of the area (hex, rgb, rgba, hsl, hsv or any CSS named colors https://www.w3.org/TR/css-color-3/#svg-color\n",
       "\n",
       "* **nbins** (default: 200) [int]\n",
       "\n",
       "Number of bins to devide the x axis in\n",
       "\n",
       "* **smooth_sigma** (default: 2) [float]\n",
       "\n",
       "standard deviation for Gaussian kernel\n",
       "\n",
       "* **width** (default: None) [int]\n",
       "\n",
       "With of the plotting area in pixel\n",
       "\n",
       "* **height** (default: 500) [int]\n",
       "\n",
       "height of the plotting area in pixel\n",
       "\n",
       "* **plot_title** (default: Basecalled reads PHRED quality) [str]\n",
       "\n",
       "Title to display on top of the plot\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(pycoQC_plot.read_qual_1D)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-09T16:48:43.495388Z",
     "start_time": "2020-01-09T16:48:43.027697Z"
    }
   },
   "outputs": [],
   "source": [
    "p = pycoQC(\"./data/Albacore-2.1.10_basecall-1D-RNA_sequencing_summary.txt.gz\", min_pass_len=200, min_pass_qual=7, quiet=True)\n",
    "p.read_qual_1D()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:24.855126Z",
     "start_time": "2020-02-10T14:23:24.848037Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**read_len_read_qual_2D** (colorscale, x_nbins, y_nbins, smooth_sigma, width, height, plot_title)\n",
       "\n",
       "Plot a 2D distribution of quality scores vs length of the reads\n",
       "\n",
       "---\n",
       "\n",
       "* **colorscale** (default: [[0.0, 'rgba(255,255,255,0)'], [0.1, 'rgba(255,150,0,0)'], [0.25, 'rgb(255,100,0)'], [0.5, 'rgb(200,0,0)'], [0.75, 'rgb(120,0,0)'], [1.0, 'rgb(70,0,0)']])\n",
       "\n",
       "a valid plotly color scale https://plot.ly/python/colorscales/ (Not recommanded to change)\n",
       "\n",
       "* **x_nbins** (default: 200) [int]\n",
       "\n",
       "Number of bins to divide the read length values in (x axis)\n",
       "\n",
       "* **y_nbins** (default: 100) [int]\n",
       "\n",
       "Number of bins to divide the read quality values in (y axis)\n",
       "\n",
       "* **smooth_sigma** (default: 2) [float]\n",
       "\n",
       "standard deviation for 2D Gaussian kernel\n",
       "\n",
       "* **width** (default: None) [int]\n",
       "\n",
       "With of the plotting area in pixel\n",
       "\n",
       "* **height** (default: 600) [int]\n",
       "\n",
       "height of the plotting area in pixel\n",
       "\n",
       "* **plot_title** (default: Basecalled reads length vs reads PHRED quality) [str]\n",
       "\n",
       "Title to display on top of the plot\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(pycoQC_plot.read_len_read_qual_2D)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-09T16:48:44.911861Z",
     "start_time": "2020-01-09T16:48:43.506873Z"
    },
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "p = pycoQC(\"./data/*Albacore*DNA*\", min_pass_len=200, min_pass_qual=7, quiet=True)\n",
    "p.read_len_read_qual_2D ()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sequencing output, quality and length over experiment time"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-11-07T15:44:50.347827Z",
     "start_time": "2018-11-07T15:44:50.340049Z"
    }
   },
   "source": [
    "pycoQC can generate plot showing the evolution of the sequencing output (`output_over_time`), the mean read quality (`qual_over_time`) and the read length (`len_over_time`) over the course of the sequencing run.\n",
    "\n",
    "Please be aware that if there are multiple run IDs in the source file(s), pycoQC reorder the run IDS by decreasing throughput/second as explained in **Initialisation**. This means that the over_time plots could be wrong, particularly when mixing several runs together.\n",
    "\n",
    "For both functions the argument `smooth_sigma` can be used to modulate the smoothing factor of the gaussian filter, if you are not satisfied with the default result.\n",
    "\n",
    "The colors of both plots can be fully customised:\n",
    "* `cumulative_color` and `interval_color` for `output_over_time`\n",
    "* `median_color`, `quartile_color` and `extreme_color` for `quality_over_time`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:24.865984Z",
     "start_time": "2020-02-10T14:23:24.858031Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**output_over_time** (cumulative_color, interval_color, time_bins, width, height, plot_title)\n",
       "\n",
       "Plot a yield over time\n",
       "\n",
       "---\n",
       "\n",
       "* **cumulative_color** (default: rgb(204,226,255)) [str]\n",
       "\n",
       "Color of cumulative yield area (hex, rgb, rgba, hsl, hsv or any CSS named colors https://www.w3.org/TR/css-color-3/#svg-color\n",
       "\n",
       "* **interval_color** (default: rgb(102,168,255)) [str]\n",
       "\n",
       "Color of interval yield line (hex, rgb, rgba, hsl, hsv or any CSS named colors https://www.w3.org/TR/css-color-3/#svg-color\n",
       "\n",
       "* **time_bins** (default: 500) [int]\n",
       "\n",
       "Number of bins to divide the time values in (x axis)\n",
       "\n",
       "* **width** (default: None) [int]\n",
       "\n",
       "With of the plotting area in pixel\n",
       "\n",
       "* **height** (default: 500) [int]\n",
       "\n",
       "height of the plotting area in pixel\n",
       "\n",
       "* **plot_title** (default: Output over experiment time) [str]\n",
       "\n",
       "Title to display on top of the plot\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(pycoQC_plot.output_over_time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-09T16:48:45.431622Z",
     "start_time": "2020-01-09T16:48:44.926359Z"
    }
   },
   "outputs": [],
   "source": [
    "p  = pycoQC (\"./data/Albacore-1.2.1_basecall-1D-DNA_sequencing_summary.txt.gz\", min_pass_len=200, min_pass_qual=7, quiet=True)\n",
    "p.output_over_time ()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:24.874714Z",
     "start_time": "2020-02-10T14:23:24.868869Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**read_qual_over_time** (median_color, quartile_color, extreme_color, smooth_sigma, time_bins, width, height, plot_title)\n",
       "\n",
       "Plot a mean quality over time\n",
       "\n",
       "---\n",
       "\n",
       "* **median_color** (default: rgb(250,128,114)) [str]\n",
       "\n",
       "Color of median line color (hex, rgb, rgba, hsl, hsv or any CSS named colors https://www.w3.org/TR/css-color-3/#svg-color\n",
       "\n",
       "* **quartile_color** (default: rgb(250,170,160)) [str]\n",
       "\n",
       "Color of inter quartile area and lines (hex, rgb, rgba, hsl, hsv or any CSS named colors https://www.w3.org/TR/css-color-3/#svg-color\n",
       "\n",
       "* **extreme_color** (default: rgba(250,170,160,0.5)) [str]\n",
       "\n",
       "Color of inter extreme area and lines (hex, rgb, rgba, hsl, hsv or any CSS named colors https://www.w3.org/TR/css-color-3/#svg-col\n",
       "\n",
       "* **smooth_sigma** (default: 1) [float]\n",
       "\n",
       "sigma parameter for the Gaussian filter line smoothing\n",
       "\n",
       "* **time_bins** (default: 500) [int]\n",
       "\n",
       "Number of bins to divide the time values in (x axis)\n",
       "\n",
       "* **width** (default: None) [int]\n",
       "\n",
       "With of the plotting area in pixel\n",
       "\n",
       "* **height** (default: 500) [int]\n",
       "\n",
       "height of the plotting area in pixel\n",
       "\n",
       "* **plot_title** (default: Read quality over experiment time) [str]\n",
       "\n",
       "Title to display on top of the plot\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp (pycoQC_plot.read_qual_over_time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-09T16:48:46.118592Z",
     "start_time": "2020-01-09T16:48:45.440431Z"
    }
   },
   "outputs": [],
   "source": [
    "p  = pycoQC (\"./data/Albacore-2.1.10_basecall-1D-DNA_sequencing_summary.txt.gz\", min_pass_len=200, min_pass_qual=7, quiet=True)\n",
    "p.read_qual_over_time ()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:24.883498Z",
     "start_time": "2020-02-10T14:23:24.877090Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**read_len_over_time** (median_color, quartile_color, extreme_color, smooth_sigma, time_bins, width, height, plot_title)\n",
       "\n",
       "Plot a read length over time\n",
       "\n",
       "---\n",
       "\n",
       "* **median_color** (default: rgb(102,168,255)) [str]\n",
       "\n",
       "Color of median line color (hex, rgb, rgba, hsl, hsv or any CSS named colors https://www.w3.org/TR/css-color-3/#svg-color\n",
       "\n",
       "* **quartile_color** (default: rgb(153,197,255)) [str]\n",
       "\n",
       "Color of inter quartile area and lines (hex, rgb, rgba, hsl, hsv or any CSS named colors https://www.w3.org/TR/css-color-3/#svg-color\n",
       "\n",
       "* **extreme_color** (default: rgba(153,197,255,0.5)) [str]\n",
       "\n",
       "Color of inter extreme area and lines (hex, rgb, rgba, hsl, hsv or any CSS named colors https://www.w3.org/TR/css-color-3/#svg-col\n",
       "\n",
       "* **smooth_sigma** (default: 1) [float]\n",
       "\n",
       "sigma parameter for the Gaussian filter line smoothing\n",
       "\n",
       "* **time_bins** (default: 500) [int]\n",
       "\n",
       "Number of bins to divide the time values in (x axis)\n",
       "\n",
       "* **width** (default: None) [int]\n",
       "\n",
       "With of the plotting area in pixel\n",
       "\n",
       "* **height** (default: 500) [int]\n",
       "\n",
       "height of the plotting area in pixel\n",
       "\n",
       "* **plot_title** (default: Read length over experiment time) [str]\n",
       "\n",
       "Title to display on top of the plot\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp (pycoQC_plot.read_len_over_time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-09T16:48:46.963715Z",
     "start_time": "2020-01-09T16:48:46.140186Z"
    }
   },
   "outputs": [],
   "source": [
    "p  = pycoQC (\"./data/Albacore-2.1.10_basecall-1D-DNA_sequencing_summary.txt.gz\", min_pass_len=200, min_pass_qual=7, quiet=True)\n",
    "p.read_len_over_time ()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Barcode distribution"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-11-07T16:05:00.048500Z",
     "start_time": "2018-11-07T16:05:00.042746Z"
    }
   },
   "source": [
    "When barcoding information is available, it is possible to generate a pie chart of the barcode count distribution. If no barcode information is available pycoQC throws an error.\n",
    "\n",
    "It is not rare to have non-relevant barcodes detected at very low level. By default any barcode below 0.1% of the reads is excludes from the plot, but this can be changed with `min_percent_barcode`.\n",
    "\n",
    "Similar to the previously described methods colors are customisable with `colors`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:24.893589Z",
     "start_time": "2020-02-10T14:23:24.885668Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**barcode_counts** (colors, width, height, plot_title)\n",
       "\n",
       "Plot a mean quality over time\n",
       "\n",
       "---\n",
       "\n",
       "* **colors** (default: ['#f8bc9c', '#f6e9a1', '#f5f8f2', '#92d9f5', '#4f97ba']) [list]\n",
       "\n",
       "List of colors (hex, rgb, rgba, hsl, hsv or any CSS named colors https://www.w3.org/TR/css-color-3/#svg-color\n",
       "\n",
       "* **width** (default: None) [int]\n",
       "\n",
       "With of the plotting area in pixel\n",
       "\n",
       "* **height** (default: 500) [int]\n",
       "\n",
       "height of the plotting area in pixel\n",
       "\n",
       "* **plot_title** (default: Percentage of reads per barcode) [str]\n",
       "\n",
       "Title to display on top of the plot\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(pycoQC_plot.barcode_counts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Albacore output example "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-09T16:48:47.335883Z",
     "start_time": "2020-01-09T16:48:46.971360Z"
    },
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "p  = pycoQC (\"./data/Albacore-1.7.0_basecall-1D-DNA_sequencing_summary.txt.gz\", min_pass_len=200, min_pass_qual=7, quiet=True)\n",
    "p.barcode_counts ()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Guppy 2.1 output example "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-09T16:48:47.942049Z",
     "start_time": "2020-01-09T16:48:47.337879Z"
    },
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "p  = pycoQC (\n",
    "    summary_file=\"./data/Guppy-2.1.3_basecall-1D-DNA_sequencing_summary.txt.gz\",\n",
    "    barcode_file=\"./data/Guppy-2.1.3_basecall-1D_DNA_barcoding_summary.txt.gz\",\n",
    "    quiet=True)\n",
    "p.barcode_counts ()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Deepbinner output example "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-09T16:48:48.382118Z",
     "start_time": "2020-01-09T16:48:47.944341Z"
    }
   },
   "outputs": [],
   "source": [
    "p  = pycoQC (\n",
    "    summary_file=\"./data/Guppy-basecall-1D-DNA_sequencing_summary.txt.gz\",\n",
    "    barcode_file=\"./data/Guppy-basecall-1D-DNA_deepbinner_barcoding_summary.txt.gz\",\n",
    "    min_pass_len=200, \n",
    "    min_pass_qual=7,\n",
    "    quiet=True)\n",
    "p.barcode_counts ()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Channels activity over time"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-11-07T16:21:54.790481Z",
     "start_time": "2018-11-07T16:21:54.782434Z"
    }
   },
   "source": [
    "Although the flowcell layout could be visually attractive (see https://github.com/mattloose/flowcellvis) this is not very informative on how the channels generate data during the run.\n",
    "\n",
    "The `channels_activity` method generates a heatmap style plot showing the output over time per channel.\n",
    "\n",
    "The number of channels can be changed to match Minion flowcells (512 default) or Promethion flowcells (3000).\n",
    "\n",
    "The argument `smooth_sigma` can be used to modulate the smoothing factor of the gaussian smoothing filter\n",
    "\n",
    "Colors can be changed with `colorscale`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:24.903049Z",
     "start_time": "2020-02-10T14:23:24.896097Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**channels_activity** (colorscale, smooth_sigma, time_bins, width, height, plot_title)\n",
       "\n",
       "Plot a yield over time\n",
       "\n",
       "---\n",
       "\n",
       "* **colorscale** (default: [[0.0, 'rgba(255,255,255,0)'], [0.01, 'rgb(255,255,200)'], [0.25, 'rgb(255,200,0)'], [0.5, 'rgb(200,0,0)'], [0.75, 'rgb(120,0,0)'], [1.0, 'rgb(0,0,0)']]) [list]\n",
       "\n",
       "a valid plotly color scale https://plot.ly/python/colorscales/ (Not recommanded to change)\n",
       "\n",
       "* **smooth_sigma** (default: 1) [float]\n",
       "\n",
       "sigma parameter for the Gaussian filter line smoothing\n",
       "\n",
       "* **time_bins** (default: 100) [int]\n",
       "\n",
       "Number of bins to divide the time values in (y axis)\n",
       "\n",
       "* **width** (default: None) [int]\n",
       "\n",
       "With of the plotting area in pixel\n",
       "\n",
       "* **height** (default: 600) [int]\n",
       "\n",
       "height of the plotting area in pixel\n",
       "\n",
       "* **plot_title** (default: Output per channel over experiment time) [str]\n",
       "\n",
       "Title to display on top of the plot\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(pycoQC_plot.channels_activity)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-09T16:48:50.777265Z",
     "start_time": "2020-01-09T16:48:48.395143Z"
    },
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "p  = pycoQC (\"./data/Albacore-2.1.10_basecall-1D-DNA_sequencing_summary.txt.gz\", quiet=True)\n",
    "p.channels_activity ()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Alignment length and score distribution"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From version 2.5+ PycoQC can also generate ditribution plots for alignments length and score and if a bam file corresponding to the summary file is provided.\n",
    "pycoQC has 5 methods related to alignment score and length:  \n",
    "\n",
    "* `align_len_1D`: A density plot of (primary) alignments length in logarithmic scale\n",
    "* `align_score_1D`: A density plot of alignments score corresponding to the invert edit distance normalised by alignment length (note that you bam file need )\n",
    "* `align_len_align_score_2D`: A density contour plot of alignments length vs alignments score in semi-log scale. \n",
    "* `read_len_align_len_2D`: A density contour plot of estimated basecalled reads length compared vs the actual alignments length in log-log scale. \n",
    "* `read_qual_align_score_2D`: A density contour plot of Read PHRED quality vs alignments score.\n",
    "\n",
    "Non-mapped reads are not represented in those plots, and attempting to use the functions without bam file will trigger an error."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:24.917110Z",
     "start_time": "2020-02-10T14:23:24.906515Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**align_len_1D** (color, nbins, smooth_sigma, width, height, plot_title)\n",
       "\n",
       "Plot a distribution of read length (log scale)\n",
       "\n",
       "---\n",
       "\n",
       "* **color** (default: mediumseagreen) [str]\n",
       "\n",
       "Color of the area (hex, rgb, rgba, hsl, hsv or any CSS named colors https://www.w3.org/TR/css-color-3/#svg-color\n",
       "\n",
       "* **nbins** (default: 200) [int]\n",
       "\n",
       "Number of bins to devide the x axis in\n",
       "\n",
       "* **smooth_sigma** (default: 2) [float]\n",
       "\n",
       "standard deviation for Gaussian kernel\n",
       "\n",
       "* **width** (default: None) [int]\n",
       "\n",
       "With of the plotting area in pixel\n",
       "\n",
       "* **height** (default: 500) [int]\n",
       "\n",
       "height of the plotting area in pixel\n",
       "\n",
       "* **plot_title** (default: Aligned reads length) [str]\n",
       "\n",
       "Title to display on top of the plot\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(pycoQC_plot.align_len_1D)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-09T16:48:52.143582Z",
     "start_time": "2020-01-09T16:48:50.788930Z"
    }
   },
   "outputs": [],
   "source": [
    "p = pycoQC(summary_file=\"./large_data/sample_1_sequencing_summary.txt\", bam_file=\"./large_data/sample_1.bam\", quiet=True, min_pass_len=200, min_pass_qual=7)\n",
    "p.align_len_1D()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:24.926180Z",
     "start_time": "2020-02-10T14:23:24.919541Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**identity_freq_1D** (color, nbins, smooth_sigma, width, height, plot_title)\n",
       "\n",
       "Plot a distribution of alignments identity\n",
       "\n",
       "---\n",
       "\n",
       "* **color** (default: sandybrown) [str]\n",
       "\n",
       "Color of the area (hex, rgb, rgba, hsl, hsv or any CSS named colors https://www.w3.org/TR/css-color-3/#svg-color\n",
       "\n",
       "* **nbins** (default: 200) [int]\n",
       "\n",
       "Number of bins to devide the x axis in\n",
       "\n",
       "* **smooth_sigma** (default: 2) [float]\n",
       "\n",
       "standard deviation for Gaussian kernel\n",
       "\n",
       "* **width** (default: None) [int]\n",
       "\n",
       "With of the plotting area in pixel\n",
       "\n",
       "* **height** (default: 500) [int]\n",
       "\n",
       "height of the plotting area in pixel\n",
       "\n",
       "* **plot_title** (default: Aligned reads identity) [str]\n",
       "\n",
       "Title to display on top of the plot\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(pycoQC_plot.identity_freq_1D)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-09T16:48:53.560476Z",
     "start_time": "2020-01-09T16:48:52.153575Z"
    }
   },
   "outputs": [],
   "source": [
    "p = pycoQC(summary_file=\"./large_data/sample_1_sequencing_summary.txt\", bam_file=\"./large_data/sample_1.bam\", quiet=True, min_pass_len=200, min_pass_qual=7)\n",
    "p.identity_freq_1D()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:24.940027Z",
     "start_time": "2020-02-10T14:23:24.932608Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**align_len_identity_freq_2D** (colorscale, x_nbins, y_nbins, smooth_sigma, width, height, plot_title)\n",
       "\n",
       "Plot a 2D distribution of alignments length vs alignments identity\n",
       "\n",
       "---\n",
       "\n",
       "* **colorscale** (default: [[0.0, 'rgba(255,255,255,0)'], [0.1, 'rgba(255,150,0,0)'], [0.25, 'rgb(255,100,0)'], [0.5, 'rgb(200,0,0)'], [0.75, 'rgb(120,0,0)'], [1.0, 'rgb(70,0,0)']])\n",
       "\n",
       "a valid plotly color scale https://plot.ly/python/colorscales/ (Not recommanded to change)\n",
       "\n",
       "* **x_nbins** (default: 200) [int]\n",
       "\n",
       "Number of bins to divide the read length values in (x axis)\n",
       "\n",
       "* **y_nbins** (default: 100) [int]\n",
       "\n",
       "Number of bins to divide the read quality values in (y axis)\n",
       "\n",
       "* **smooth_sigma** (default: 2) [float]\n",
       "\n",
       "standard deviation for 2D Gaussian kernel\n",
       "\n",
       "* **width** (default: None) [int]\n",
       "\n",
       "With of the plotting area in pixel\n",
       "\n",
       "* **height** (default: 600) [int]\n",
       "\n",
       "height of the plotting area in pixel\n",
       "\n",
       "* **plot_title** (default: Aligned reads length vs alignments identity) [str]\n",
       "\n",
       "Title to display on top of the plot\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(pycoQC_plot.align_len_identity_freq_2D)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-09T16:48:55.399773Z",
     "start_time": "2020-01-09T16:48:53.569405Z"
    }
   },
   "outputs": [],
   "source": [
    "p = pycoQC(summary_file=\"./large_data/sample_1_sequencing_summary.txt\", bam_file=\"./large_data/sample_1.bam\", quiet=True)\n",
    "p.align_len_identity_freq_2D()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:24.954614Z",
     "start_time": "2020-02-10T14:23:24.942505Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**read_len_align_len_2D** (colorscale, x_nbins, y_nbins, smooth_sigma, width, height, plot_title)\n",
       "\n",
       "Plot a 2D distribution of length of the reads vs length of the alignments\n",
       "\n",
       "---\n",
       "\n",
       "* **colorscale** (default: [[0.0, 'rgba(255,255,255,0)'], [0.1, 'rgba(255,150,0,0)'], [0.25, 'rgb(255,100,0)'], [0.5, 'rgb(200,0,0)'], [0.75, 'rgb(120,0,0)'], [1.0, 'rgb(70,0,0)']])\n",
       "\n",
       "a valid plotly color scale https://plot.ly/python/colorscales/ (Not recommanded to change)\n",
       "\n",
       "* **x_nbins** (default: 200) [int]\n",
       "\n",
       "Number of bins to divide the read length values in (x axis)\n",
       "\n",
       "* **y_nbins** (default: 100) [int]\n",
       "\n",
       "Number of bins to divide the read quality values in (y axis)\n",
       "\n",
       "* **smooth_sigma** (default: 1) [float]\n",
       "\n",
       "standard deviation for 2D Gaussian kernel\n",
       "\n",
       "* **width** (default: None) [int]\n",
       "\n",
       "With of the plotting area in pixel\n",
       "\n",
       "* **height** (default: 600) [int]\n",
       "\n",
       "height of the plotting area in pixel\n",
       "\n",
       "* **plot_title** (default: Basecalled reads length vs alignments length) [str]\n",
       "\n",
       "Title to display on top of the plot\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(pycoQC_plot.read_len_align_len_2D)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-09T16:48:57.456151Z",
     "start_time": "2020-01-09T16:48:55.408030Z"
    }
   },
   "outputs": [],
   "source": [
    "p = pycoQC(summary_file=\"./large_data/sample_1_sequencing_summary.txt\", bam_file=\"./large_data/sample_1.bam\", quiet=True)\n",
    "p.read_len_align_len_2D()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:24.965071Z",
     "start_time": "2020-02-10T14:23:24.958066Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**read_qual_identity_freq_2D** (colorscale, x_nbins, y_nbins, smooth_sigma, width, height, plot_title)\n",
       "\n",
       "Plot a 2D distribution of read quality vs alignments identity\n",
       "\n",
       "---\n",
       "\n",
       "* **colorscale** (default: [[0.0, 'rgba(255,255,255,0)'], [0.1, 'rgba(255,150,0,0)'], [0.25, 'rgb(255,100,0)'], [0.5, 'rgb(200,0,0)'], [0.75, 'rgb(120,0,0)'], [1.0, 'rgb(70,0,0)']])\n",
       "\n",
       "a valid plotly color scale https://plot.ly/python/colorscales/ (Not recommanded to change)\n",
       "\n",
       "* **x_nbins** (default: 200) [int]\n",
       "\n",
       "Number of bins to divide the read length values in (x axis)\n",
       "\n",
       "* **y_nbins** (default: 100) [int]\n",
       "\n",
       "Number of bins to divide the read quality values in (y axis)\n",
       "\n",
       "* **smooth_sigma** (default: 1) [float]\n",
       "\n",
       "standard deviation for 2D Gaussian kernel\n",
       "\n",
       "* **width** (default: None) [int]\n",
       "\n",
       "With of the plotting area in pixel\n",
       "\n",
       "* **height** (default: 600) [int]\n",
       "\n",
       "height of the plotting area in pixel\n",
       "\n",
       "* **plot_title** (default: Reads PHRED quality vs alignments identity) [str]\n",
       "\n",
       "Title to display on top of the plot\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(pycoQC_plot.read_qual_identity_freq_2D)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-09T16:48:59.484500Z",
     "start_time": "2020-01-09T16:48:57.471730Z"
    },
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "p = pycoQC(summary_file=\"./large_data/sample_1_sequencing_summary.txt\", bam_file=\"./large_data/sample_1.bam\", quiet=True, min_pass_len=200, min_pass_qual=7)\n",
    "p.read_qual_identity_freq_2D()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:24.976088Z",
     "start_time": "2020-02-10T14:23:24.967858Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**align_len_over_time** (median_color, quartile_color, extreme_color, smooth_sigma, time_bins, width, height, plot_title)\n",
       "\n",
       "Plot a aligned reads length over time\n",
       "\n",
       "---\n",
       "\n",
       "* **median_color** (default: rgb(102,168,255)) [str]\n",
       "\n",
       "Color of median line color (hex, rgb, rgba, hsl, hsv or any CSS named colors https://www.w3.org/TR/css-color-3/#svg-color\n",
       "\n",
       "* **quartile_color** (default: rgb(153,197,255)) [str]\n",
       "\n",
       "Color of inter quartile area and lines (hex, rgb, rgba, hsl, hsv or any CSS named colors https://www.w3.org/TR/css-color-3/#svg-color\n",
       "\n",
       "* **extreme_color** (default: rgba(153,197,255,0.5)) [str]\n",
       "\n",
       "Color of inter extreme area and lines (hex, rgb, rgba, hsl, hsv or any CSS named colors https://www.w3.org/TR/css-color-3/#svg-col\n",
       "\n",
       "* **smooth_sigma** (default: 1) [float]\n",
       "\n",
       "sigma parameter for the Gaussian filter line smoothing\n",
       "\n",
       "* **time_bins** (default: 500) [int]\n",
       "\n",
       "Number of bins to divide the time values in (x axis)\n",
       "\n",
       "* **width** (default: None) [int]\n",
       "\n",
       "With of the plotting area in pixel\n",
       "\n",
       "* **height** (default: 500) [int]\n",
       "\n",
       "height of the plotting area in pixel\n",
       "\n",
       "* **plot_title** (default: Aligned reads length over experiment time) [str]\n",
       "\n",
       "Title to display on top of the plot\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(pycoQC_plot.align_len_over_time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-09T16:49:00.972507Z",
     "start_time": "2020-01-09T16:48:59.499038Z"
    }
   },
   "outputs": [],
   "source": [
    "p = pycoQC(summary_file=\"./large_data/sample_1_sequencing_summary.txt\", bam_file=\"./large_data/sample_1.bam\", quiet=True, min_pass_len=200, min_pass_qual=7)\n",
    "p.align_len_over_time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:24.985866Z",
     "start_time": "2020-02-10T14:23:24.979275Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**identity_freq_over_time** (median_color, quartile_color, extreme_color, smooth_sigma, time_bins, width, height, plot_title)\n",
       "\n",
       "Plot the alignment identity scores over time\n",
       "\n",
       "---\n",
       "\n",
       "* **median_color** (default: rgb(250,128,114)) [str]\n",
       "\n",
       "Color of median line color (hex, rgb, rgba, hsl, hsv or any CSS named colors https://www.w3.org/TR/css-color-3/#svg-color\n",
       "\n",
       "* **quartile_color** (default: rgb(250,170,160)) [str]\n",
       "\n",
       "Color of inter quartile area and lines (hex, rgb, rgba, hsl, hsv or any CSS named colors https://www.w3.org/TR/css-color-3/#svg-color\n",
       "\n",
       "* **extreme_color** (default: rgba(250,170,160,0.5)) [str]\n",
       "\n",
       "Color of inter extreme area and lines (hex, rgb, rgba, hsl, hsv or any CSS named colors https://www.w3.org/TR/css-color-3/#svg-col\n",
       "\n",
       "* **smooth_sigma** (default: 1) [float]\n",
       "\n",
       "sigma parameter for the Gaussian filter line smoothing\n",
       "\n",
       "* **time_bins** (default: 500) [int]\n",
       "\n",
       "Number of bins to divide the time values in (x axis)\n",
       "\n",
       "* **width** (default: None) [int]\n",
       "\n",
       "With of the plotting area in pixel\n",
       "\n",
       "* **height** (default: 500) [int]\n",
       "\n",
       "height of the plotting area in pixel\n",
       "\n",
       "* **plot_title** (default: Aligned reads identity over experiment time) [str]\n",
       "\n",
       "Title to display on top of the plot\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(pycoQC_plot.identity_freq_over_time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-09T16:49:03.017620Z",
     "start_time": "2020-01-09T16:49:00.979474Z"
    }
   },
   "outputs": [],
   "source": [
    "p = pycoQC(summary_file=\"./large_data/sample_1_sequencing_summary.txt\", bam_file=\"./large_data/sample_1.bam\", quiet=True, min_pass_len=200, min_pass_qual=7)\n",
    "p.identity_freq_over_time()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Alignments summaries and coverage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If a bam is provided pycoQC can plot 3 alignments specific plots:\n",
    "* `alignment_rate`: Generates a table and a Sankey Diagram of the number of bases basecalled, mapped, matching...\n",
    "* `alignment_summary`: Generates a table and a Pie plot indicating primary, secondary, supplementary and unmapped reads \n",
    "* `alignment_coverage`: Generates a simple coverage plot overview over all the target genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:24.997018Z",
     "start_time": "2020-02-10T14:23:24.989242Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**alignment_rate** (colors, width, height, plot_title)\n",
       "\n",
       "Plot a basic alignment summary\n",
       "\n",
       "---\n",
       "\n",
       "* **colors** (default: ['#fcaf94', '#828282', '#fc8161', '#828282', '#f44f39', '#d52221', '#828282', '#828282', '#828282', '#828282']) [list]\n",
       "\n",
       "List of colors (hex, rgb, rgba, hsl, hsv or any CSS named colors https://www.w3.org/TR/css-color-3/#svg-color\n",
       "\n",
       "* **width** (default: None) [int]\n",
       "\n",
       "With of the plotting area in pixel\n",
       "\n",
       "* **height** (default: 600) [int]\n",
       "\n",
       "height of the plotting area in pixel\n",
       "\n",
       "* **plot_title** (default: Bases alignment rate) [str]\n",
       "\n",
       "Title to display on top of the plot\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(pycoQC_plot.alignment_rate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-09T16:49:04.611877Z",
     "start_time": "2020-01-09T16:49:03.023740Z"
    }
   },
   "outputs": [],
   "source": [
    "p = pycoQC(summary_file=\"./large_data/sample_2_sequencing_summary.txt\", bam_file=\"./large_data/sample_2.bam\", quiet=True)\n",
    "p.alignment_rate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:25.006147Z",
     "start_time": "2020-02-10T14:23:24.999860Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**alignment_reads_status** (colors, width, height, plot_title)\n",
       "\n",
       "Plot a basic alignment summary\n",
       "\n",
       "---\n",
       "\n",
       "* **colors** (default: ['#f44f39', '#fc8161', '#fcaf94', '#828282']) [list]\n",
       "\n",
       "List of colors (hex, rgb, rgba, hsl, hsv or any CSS named colors https://www.w3.org/TR/css-color-3/#svg-color\n",
       "\n",
       "* **width** (default: None) [int]\n",
       "\n",
       "With of the plotting area in pixel\n",
       "\n",
       "* **height** (default: 500) [int]\n",
       "\n",
       "height of the plotting area in pixel\n",
       "\n",
       "* **plot_title** (default: Summary of reads alignment status) [str]\n",
       "\n",
       "Title to display on top of the plot\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(pycoQC_plot.alignment_reads_status)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-09T16:49:06.151721Z",
     "start_time": "2020-01-09T16:49:04.635774Z"
    }
   },
   "outputs": [],
   "source": [
    "p = pycoQC(summary_file=\"./large_data/sample_2_sequencing_summary.txt\", bam_file=\"./large_data/sample_2.bam\", quiet=True)\n",
    "p.alignment_reads_status()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:23:25.014913Z",
     "start_time": "2020-02-10T14:23:25.008698Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**alignment_coverage** (nbins, color, smooth_sigma, width, height, plot_title)\n",
       "\n",
       "Plot coverage over all the references\n",
       "\n",
       "---\n",
       "\n",
       "* **nbins** (default: 500) [int]\n",
       "\n",
       "Number of bins to divide the coverage into.\n",
       "\n",
       "* **color** (default: rgba(70,130,180,0.70)) [str]\n",
       "\n",
       "* **smooth_sigma** (default: 1) [int]\n",
       "\n",
       "sigma parameter for the Gaussian filter line smoothing\n",
       "\n",
       "* **width** (default: None) [int]\n",
       "\n",
       "With of the plotting area in pixel\n",
       "\n",
       "* **height** (default: 500) [int]\n",
       "\n",
       "height of the plotting area in pixel\n",
       "\n",
       "* **plot_title** (default: Coverage overview) [str]\n",
       "\n",
       "Title to display on top of the plot\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(pycoQC_plot.alignment_coverage)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-09T16:49:07.724347Z",
     "start_time": "2020-01-09T16:49:06.160984Z"
    }
   },
   "outputs": [],
   "source": [
    "p = pycoQC(summary_file=\"./large_data/sample_2_sequencing_summary.txt\", bam_file=\"./large_data/sample_2.bam\", quiet=True)\n",
    "p.alignment_coverage()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generating HTML and json reports"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From version 2.3+ the PycoQC API can also generate an interactive HTML report and a text JSON report similar to the CLI functionality."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:31:30.374597Z",
     "start_time": "2020-02-10T14:31:25.170948Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Checking arguments values\n",
      "Check input data files\n",
      "Parse data files\n",
      "Merge data\n",
      "Cleaning data\n",
      "\tDiscarding lines containing NA values\n",
      "\t\t0 reads discarded\n",
      "\tFiltering out zero length reads\n",
      "\t\t438 reads discarded\n",
      "\tSorting run IDs by decreasing throughput\n",
      "\t\tRun-id order ['db5916f2fe7957afac1d0aaccdec883342c4bc31']\n",
      "\tReordering runids\n",
      "\t\tProcessing reads with Run_ID db5916f2fe7957afac1d0aaccdec883342c4bc31 / time offset: 0\n",
      "\tCleaning up low frequency barcodes\n",
      "\t\t0 reads with low frequency barcode unset\n",
      "\tCast value to appropriate type\n",
      "\tReindexing dataframe by read_ids\n",
      "\t\t9,562 Final valid reads\n",
      "Loading plotting interface\n",
      "\tFound 9,562 total reads\n",
      "\tFound 8,042 pass reads (qual >= 7.0 and length >= 0)\n",
      "Generating HTML report\n",
      "\tParsing html config file\n",
      "\tRunning method run_summary\n",
      "\t\tComputing plot\n",
      "\tRunning method basecall_summary\n",
      "\t\tComputing plot\n",
      "\tRunning method alignment_summary\n",
      "\t\tNo Alignment information available\n",
      "\tRunning method read_len_1D\n",
      "\t\tComputing plot\n",
      "\tRunning method align_len_1D\n",
      "\t\tNo Alignment information available\n",
      "\tRunning method read_qual_1D\n",
      "\t\tComputing plot\n",
      "\tRunning method identity_freq_1D\n",
      "\t\tNo identity frequency information available\n",
      "\tRunning method read_len_read_qual_2D\n",
      "\t\tComputing plot\n",
      "\tRunning method read_len_align_len_2D\n",
      "\t\tNo Alignment information available\n",
      "\tRunning method align_len_identity_freq_2D\n",
      "\t\tNo identity frequency information available\n",
      "\tRunning method read_qual_identity_freq_2D\n",
      "\t\tNo identity frequency information available\n",
      "\tRunning method output_over_time\n",
      "\t\tComputing plot\n",
      "\tRunning method read_len_over_time\n",
      "\t\tComputing plot\n",
      "\tRunning method read_qual_over_time\n",
      "\t\tComputing plot\n",
      "\tRunning method align_len_over_time\n",
      "\t\tNo Alignment information available\n",
      "\tRunning method identity_freq_over_time\n",
      "\t\tNo identity frequency information available\n",
      "\tRunning method barcode_counts\n",
      "\t\tComputing plot\n",
      "\tRunning method channels_activity\n",
      "\t\tComputing plot\n",
      "\tRunning method alignment_reads_status\n",
      "\t\tNo Alignment information available\n",
      "\tRunning method alignment_rate\n",
      "\t\tNo identity frequency information available\n",
      "\tRunning method alignment_coverage\n",
      "\t\tNo Alignment information available\n",
      "\tLoading HTML template\n",
      "\tRendering plots in d3js\n",
      "\tWriting to HTML file\n",
      "Generating JSON report\n",
      "\tRunning summary_stats_dict method\n",
      "\tCompute overall summary statistics\n",
      "\tWriting to JSON file\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tBarcode: True\n",
      "\tAlignment: False\n",
      "\tPromethion: False\n",
      "\tAll reads: 9,562\n",
      "\tAll bases: 14,664,334\n",
      "\tAll median read length: 688.5\n",
      "\tPass reads: 8,042\n",
      "\tPass bases: 12,823,819\n",
      "\tPass median read length: 716.0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "p = pycoQC(\n",
    "    summary_file=\"./data/Albacore-1.7.0_basecall-1D-DNA_sequencing_summary.txt.gz\",\n",
    "    html_outfile=\"./results/Albacore-1.7.0_basecall-1D-DNA_API.html\",\n",
    "    json_outfile=\"./results/Albacore-1.7.0_basecall-1D-DNA_API.json\")\n",
    "\n",
    "print (p)"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Initialization Cell",
  "hide_input": false,
  "kernelspec": {
   "display_name": "pycoQC",
   "language": "python",
   "name": "pycoqc"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  },
  "widgets": {
   "state": {},
   "version": "1.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
