---
title: 'pycoQC, interactive quality control for Oxford Nanopore Sequencing '
tags:
  - python3
  - plots
  - metrics
  - nanopore 
  - interactive
authors:
  - name: Adrien Leger
    orcid: 0000-0001-8772-6776
    affiliation: 1
  -name: Tommaso Leonardi
    orcid: 0000-0002-4449-1863
    affiliation: "2,3"
affiliations:
 - name: EMBL-EBI, Hinxton, Cambridgeshire
   index: 1
 - name: Gurdon Institute, Cambridge, Cambridgeshire
   index: 2
 - name: Center for Genomic Science IIT@SEMM, Istituto Italiano di Tecnologia (IIT), Milan, Italy
   index: 3
date: 08 November 2018
bibliography: paper.bib
---

# Summary

Nanopore sequencing of nucleic acids took nearly 30 years to develop and is now firmly established as a reliable alternative to sequencing by synthesis methods [@deamer_three_2016]. Oxford Nanopore Technologies (ONT) released the first commercial nanopore sequencing device in 2014 and has continually improved the technology since then [@jain_oxford_2016]. Although the read accuracy is only around 90%, ONT technology can sequence very long molecules and generates data in real time. In addition, RNA can be sequenced directly and modified bases can be detected [@garalde2018highly].

The electrical signal acquired by the array of nanopores is stored in hdf5 format, with one file (called fast5) per molecule sequenced. The signal is then converted into a nucleic acid sequence using a basecalling software. There are several alternatives, but the best performers for read accuracy are Albacore and Guppy developed and maintained by ONT [@ryan_wick_comparison_2018]. Both softwares can generate fastq files, fast5 files containing basecalling information and a text summary file. Although ONT released best-practice guidelines for quality control analysis of  sequencing run [@oxford_nanopore_technologies_bioinformatics_2019], it did not provide any streamlined quality control software to evaluate in depth the quality of the sequencing run.

Here we present pycoQC, a new tool to generate interactive quality control metrics and plots from basecalled nanopore reads or from the summary file generated by Albacore and Guppy. Although there are other open-source alternatives such as Nanoplot [@de_coster_nanopack_2018], MinionQC [@lanfear_minionqc_2018] and toulligQC [@berengere_laffay_toulligqc_2018], pycoQC has several novel unique features:

* Integration with plotly [@plotly_technologies_inc_plotly_2015] python charting library to create dynamic D3.js data visualizations.
* Extensive python API developed for interactive data exploration in Jupyter Notebook [@jupyter_project_and_community_jupyter_2019] ([example notebook](https://a-slide.github.io/pycoQC/pycoQC_usage.html)).
* Simple command line interface to generate customizable interactive HTML reports ([example report](https://a-slide.github.io/pycoQC/pycoQC_Albacore-2.1.10_basecall-1D-DNA.htmll)).
* Multiprocessing Fast5 feature extraction program to generate a summary file directly from Fast5 files if not available.
* Support for data generated by ONT MinION, GridION and PromethION devices basecalled by Albacore 1.3+, Guppy 2.1.3+ or MinKNOW 18.12+. 

# Principle and example output

Briefly, pycoQC imports, filters and preprocesses one or several summary files generated with either one of the previously mentioned basecallers or with the companion program `Fast5_to_seq_summary` included with the package. If available, calibration strand and barcoding information are also extracted either from the summary file (Albacore) or from a separate barcoding summary (Guppy). Then, a range of plots and tables can be generated to explore the data. pycoQC plots are interactive, allowing users to display all the reads or only those above the quality threshold, to zoom in and to hide legend labels. While the command line interace offers a simple and straightforward experience, the python API for Jupyter notebook gives more flexibility to users who can easily customise and share their analyses. Static versions of some of the tables and plots produced by the software are presented in Figure 1 to 4.

![](/home/aleg/Programming/pycoQC_paper/pictures/summary.png)

**Figure 1)** Summary statistics about the run. If the summary file contains multiple run IDs then a breakdown per run ID is displayed. 



![](/home/aleg/Programming/pycoQC_paper/pictures/reads_qual_len_2D_example.png)

**Figure 2)** 2 dimensions density plot of the reads length compared with the means read PHRED quality. 1 dimension density plots of both reads quality and length  

   ![](/home/aleg/Programming/pycoQC_paper/pictures/output_over_time.png)

**Figure 3)** Read output over time

![](/home/aleg/Programming/pycoQC_paper/pictures/channels_activity.png)

**Figure 4)** Individual channel activity over time.

# Availability

pycoQC is available at https://github.com/a-slide/pycoQC together with extensive documentation. The source code has been archived on Zenodo with the linked DOI: [10.5281/zenodo.2540931](https://zenodo.org/record/2540931)

# Acknowledgements

The authors would like to thank Paulo Amaral, who generated most of the datasets used to test pycoQC as well as Tomas Fitzgerald, Jack Monahan and Michael Clark who beta-tested the package and suggested new features. In addition we would also like to thanks Kim Judge for providing us with extra summary sequencing files. A.L is funded by the ERC Cofund Marie Curie program T.L is funded by .... 

# References
